<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN">
<html>
  <head>
    <meta content="text/html; charset=utf-8" http-equiv="content-type">
    <title>RNAstructure Command Line Help: partition</title>
    <link href="Stylesheet.css" rel="stylesheet" type="text/css">
  </head>
  <body>
    <div class="bodydiv">
      <table class="mainTable" border="0" cellpadding="0" cellspacing="0">
        <tbody>
          <tr class="headerRow">
            <td class="headerBox"><img src="icon.jpg" alt="RNAstructure logo"></td>
            <td class="headerLabel">
              <h2>RNAstructure Command Line Help<br>
                Rsample and Rsample-smp</h2>
            </td>
            <td class="headerBox2">
              <ul id="navigator">
                <li>
                  <h3 class="noMarginOrPad"><a href="index.html">Contents</a></h3>
                </li>
                <li>
                  <h3 class="noMarginOrPad"><a href="AlphabeticalIndex.html">Index</a></h3>
                </li>
              </ul>
            </td>
          </tr>
          <tr>
            <td class="mainContent" colspan="3">Rsample is used to calculate the
              partition function for a sequence that has multiple conformations.
              To perform a complete Rsample calculation, the following steps
              should be performed: <br>
              <ol>
                <li>Run Rsample,  using the options listed below, to produce a <strong>Partition Save File (PFS)</strong></li>
                <li>Run 
                  <a href="stochastic.html">stochastic</a>, using this PFS file as input, to produce a <a href="File_Formats.html#CT">CT
                    file</a> with Boltzmann ensemble of 1,000 structures.</li>
                <li>Read the CT file from step 2 (as a command line argument in
                  Linux and MacOS) with <a href="https://www.r-project.org/" class="specialName">R</a> script <a href="resources/RsampleCluster.R">RsampleCluster.R</a> (optionally depends on <a href="resources/calcdistf90.R">calcdistf90.R</a> and <a href="resources/calcdist.f90">calcdist.f90</a>).
                  This program uses the algorithm by Ding and Lawrence to
                  calculate optimal number of clusters and their
                  centroids. In addition to <span class="specialName">R</span>, it requires the installation of <span class="specialName">R</span> package with clustering procedures called <span class="specialName">fpc</span> which can be
                  done by typing <span class="codeText">install.packages("fpc")</span> inside <span class="specialName">R</span>.</li>
              </ol>
              Rsample-smp is a parallel processing version for use on multi-core
              computers, built using OpenMP.<a name="Usage" id="Usage"></a>
              <h3 class="western">USAGE: Rsample &lt;seq file&gt; &lt;restraints
                file&gt; &lt;pfs file&gt; [options]</h3>
              <h3 class="western">OR: Rsample-smp &lt;seq file&gt;
                &lt;restraints file&gt; &lt;pfs file&gt; [options]</h3>
              <h3 class="leftHeader">Required parameters:</h3>
              <table style="width: 598px; height: 70px;">
                <tbody>
                  <tr>
                    <td style="width: 106.95px;" class="cell">&lt;seq file&gt;</td>
                    <td style="width: 481.05px;"> The name of a <a href="File_Formats.html#Sequence">sequence
                        file</a> containing input data. <br>
                      Note that lowercase nucleotides are forced single-stranded
                      in structure prediction. </td>
                  </tr>
                  <tr>
                    <td>&lt;restraints file&gt;</td>
                    <td>Specify a file with experimental restraints. <a href="File_Formats.html#SHAPE">SHAPE
                        data file</a> format should be used.</td>
                  </tr>
                  <tr>
                    <td class="cell">&lt;pfs file&gt;</td>
                    <td>The name of a binary partition function save file to
                      which output will be written.</td>
                  </tr>
                </tbody>
              </table>
              <h3 class="leftHeader">Options that do not require added values:</h3>
              <table>
                <tbody>
                  
                  <tr>
                    <td class="cell">--DMS</td>
                    <td>Specify that the data are DMS, not SHAPE (by default). DMS data are calibrated to in vivo DMS-MaP experiments with SSU rRNA. See Kumar et al.</td>
                  </tr>
                  <tr>
                    <td class="cell">-d, -D, --DNA</td>
                    <td> Specify that the sequence is DNA, and DNA parameters
                      are to be used. <br>
                      Default is to use RNA parameters. </td>
                  </tr>
                  <tr>
                    <td class="cell">-h, -H, --help</td>
                    <td>Display the usage details message.</td>
                  </tr>
                  <tr>
                    <td>-v, -V, --version</td>
                    <td>Display version and copyright information for this
                      interface.</td>
                  </tr>
                </tbody>
              </table>
              <h3 class="leftHeader">Options that require added values:</h3>
              <table>
                <tbody>
                <tr>
                    <td class="cell">-a, -A, --alphabet</td>
                    <td>Specify the name of a folding alphabet and associated
                      nearest neighbor parameters. The alphabet is the prefix
                      for the thermodynamic parameter files, e.g. "rna" for RNA
                      parameters or "dna" for DNA parameters or a custom
                      extended/modified alphabet. The thermodynamic parameters
                      need to reside in the at the location indicated by
                      environment variable DATAPATH.<br>
                      The default is "rna" (i.e. use RNA parameters). This
                      option overrides the --DNA flag.</td>
                  </tr>
                  <tr>
                    <td class="cell">-c, --cparam</td>
                    <td>Specify a C parameter used in Rsample calculations.<br>
                      Default value derived for SHAPE experiments is 0.5
                      kcal/mol.</td>
                  </tr>
                  <tr>
                    <td class="cell">--MAX</td>
                    <td>Specify a maximum reactivity value for data. This is 1000 by default. Using in vivo DMS-MaP experiments, we found that a value of 5 improved secondary structure prediction accuracy. The default (effectively no maximum) should work well for SHAPE data.</td>
                  </tr>
                  <tr>
                    <td class="cell">-md, --maxdistance</td>
                    <td> Specify a maximum paired distance between nucleotides.<br>
                      Default is no restriction on distance between pairs.<br>
                    </td>
                  </tr>
                   <tr>
                    <td class="cell">-ns, --numsamples</td>
                    <td> Specify number of samples for stochastic sampling
                      calculation used in Rsample.<br>
                      Default is 10,000.</td>
                  </tr>
                  <tr>
                    <td class="cell">-O, --offset</td>
                    <td> Specify and offset parameter used in Rsample
                      calculations.<br>
                      Default value derived for SHAPE experiments is 1.1
                      kcal/mol. </td>
                  </tr>
                 <tr>
                    <td class="cell">-rPE, --reactPairedEnd</td>
                    <td> Give full path to file with end-of-helix paired
                      nucleotide reactvities dataset.<br>
                      Default values (from SHAPE experiments) are in rsample
                      directory in DATAPATH. </td>
                  </tr>
                  <tr>
                    <td class="cell">-rPM, --reactPairedMiddle</td>
                    <td> Give full path to file with middle-of-helix paired
                      nucleotide reactvities dataset.<br>
                      Default values (from SHAPE experiments) are in rsample
                      directory in DATAPATH. </td>
                  </tr> 
                  <tr>
                    <td class="cell">-rUP, --reactUnpaired</td>
                    <td> Give full path to file with unpaired nucleotide
                      reactvities dataset.<br>
                      Default values (from SHAPE experiments) are in rsample
                      directory in DATAPATH.</td>
                  </tr>
                  <tr>
                    <td class="cell">-s, --seed</td>
                    <td> Specify a random seed. <br>
                      Default is to set random seed from current time. </td>
                  </tr>
                  
                  
                 
                  <tr>
                    <td class="cell">-t, -T, --temperature</td>
                    <td> Specify the temperature at which calculation takes
                      place in Kelvin. <br>
                      Default is 310.15 K, which is 37 degrees C. </td>
                  </tr>
                  
                </tbody>
              </table>
              <h3 class="leftHeader">Notes for smp:</h3>
              <p>Rsample-smp, by default, will use all available compute cores
                for processing. The number of cores used can be controlled by
                setting the OMP_NUM_THREADS environment variable.</p>
              <h3 class="leftHeader"><a name="Refs" id="Refs">References</a></h3>
              <ol class="noListTopMargin">
                <li class="gapListItem"> Reuter, J.S. and Mathews, D.H. <br>
                  "RNAstructure: software for RNA secondary structure prediction
                  and analysis." <br>
                  <i>BMC Bioinformatics</i>, 11:129. (2010). </li>
                <li class="gapListItem"> Spasic, A., Assmann, S.M., Bevilacqua,
                  P.C. and Mathews, D.H. <br>
                  "Modeling RNA secondary structure folding ensemble using SHAPE
                  mapping data." <br>
                  Nucleic Acids Research, 46: 4883 (2018).</li>
                <li class="gapListItem">Kumar, J., Lackey, L., Waldern, J.M., Dey, A., Mathews, D.H., and Laederach, A.<br>
                  &quot;Quantitative integration of RNA structure and splicing elements to explain alternative splicing of Microtubule-Associated Protein Tau gene&quot;<br>
                  In Preparation.
                </li>
                <li class="gapListItem"> Ding, Y. and Lawrence, C.E. <br>
                  "A statistical sampling algorithm for RNA secondary structure
                  prediction."&nbsp; <i></i><br>
                  <i>Nucleic Acids Research</i>, 31:7280 (2003). </li>
              </ol>
              <ol class="noListTopMargin">
              </ol>
            </td>
          </tr>
          <tr>
            <td class="footer" colspan="3">
              <p>Visit <a href="http://rna.urmc.rochester.edu/RNAstructure.html">The
                  Mathews Lab RNAstructure Page</a> for updates and latest
                information.</p>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>
</html>
