<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.13"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>RNAstructure Classes: RNA Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectlogo"><img alt="Logo" src="icon.jpg"/></td>
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">RNAstructure Classes
   &#160;<span id="projectnumber">Version 6.2</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.13 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
</script>
<div id="main-nav"></div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
$(document).ready(function(){initNavTree('class_r_n_a.html','');});
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-static-methods">Static Public Member Functions</a> &#124;
<a href="#pro-methods">Protected Member Functions</a> &#124;
<a href="#pro-attribs">Protected Attributes</a> &#124;
<a href="#pri-attribs">Private Attributes</a> &#124;
<a href="class_r_n_a-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">RNA Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p><code>#include &lt;RNA.h&gt;</code></p>
<div class="dynheader">
Inheritance diagram for RNA:</div>
<div class="dyncontent">
 <div class="center">
  <img src="class_r_n_a.png" usemap="#RNA_map" alt=""/>
  <map id="RNA_map" name="RNA_map">
<area href="class_thermodynamics.html" title="Thermodynamics Class. " alt="Thermodynamics" shape="rect" coords="177,0,285,24"/>
<area href="classdesign.html" alt="design" shape="rect" coords="0,112,108,136"/>
<area href="class_hybrid_r_n_a.html" title="HybridRNA Class. " alt="HybridRNA" shape="rect" coords="118,112,226,136"/>
<area href="class_oligowalk__object.html" title="Oligowalk_object Class. " alt="Oligowalk_object" shape="rect" coords="236,112,344,136"/>
<area href="class_prob_scan.html" alt="ProbScan" shape="rect" coords="354,112,462,136"/>
</map>
 </div></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a43195bcf410f9c53a0fb146fe490b6ea"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a43195bcf410f9c53a0fb146fe490b6ea">RNA</a> (const char sequence[], const bool IsRNA=true)</td></tr>
<tr class="memdesc:a43195bcf410f9c53a0fb146fe490b6ea"><td class="mdescLeft">&#160;</td><td class="mdescRight">Constructor - user provides a sequence as a c string.  <a href="#a43195bcf410f9c53a0fb146fe490b6ea">More...</a><br /></td></tr>
<tr class="separator:a43195bcf410f9c53a0fb146fe490b6ea"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a92d94cbac25f6e96074c209ba9fa491b"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a92d94cbac25f6e96074c209ba9fa491b">RNA</a> (const char filepathOrSequence[], const <a class="el" href="_r_n_a_8h.html#a3883db120ec3ca0fe72f322a76dd2378">RNAInputType</a> fileType, const char *const alphabetName, const bool allowUnknownBases=false, const bool <a class="el" href="class_thermodynamics.html#a018762c28a32e0a743b3a0ed5b808308">skipThermoTables</a>=false)</td></tr>
<tr class="separator:a92d94cbac25f6e96074c209ba9fa491b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9e572b36b07e0b2fee3165930a430472"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a9e572b36b07e0b2fee3165930a430472">RNA</a> (const char filepathOrSequence[], const <a class="el" href="_r_n_a_8h.html#a3883db120ec3ca0fe72f322a76dd2378">RNAInputType</a> fileType, const <a class="el" href="class_thermodynamics.html">Thermodynamics</a> *copyThermo)</td></tr>
<tr class="separator:a9e572b36b07e0b2fee3165930a430472"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a800274d25e5fe015b5672c6e7ed8214e"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a800274d25e5fe015b5672c6e7ed8214e">RNA</a> (const char filename[], const <a class="el" href="_r_n_a_8h.html#a3883db120ec3ca0fe72f322a76dd2378">RNAInputType</a> fileType, const bool IsRNA=true)</td></tr>
<tr class="memdesc:a800274d25e5fe015b5672c6e7ed8214e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Constructor - user provides a filename for existing file as a c string.  <a href="#a800274d25e5fe015b5672c6e7ed8214e">More...</a><br /></td></tr>
<tr class="separator:a800274d25e5fe015b5672c6e7ed8214e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa1ab0c1a0065091f8fdf41c2cdfb43ea"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#aa1ab0c1a0065091f8fdf41c2cdfb43ea">RNA</a> (const bool IsRNA=true)</td></tr>
<tr class="separator:aa1ab0c1a0065091f8fdf41c2cdfb43ea"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a08a3be93861700dce517caf7ccf77358"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a08a3be93861700dce517caf7ccf77358">GetErrorCode</a> () const</td></tr>
<tr class="memdesc:a08a3be93861700dce517caf7ccf77358"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return an error code, where a return of zero is no error.  <a href="#a08a3be93861700dce517caf7ccf77358">More...</a><br /></td></tr>
<tr class="separator:a08a3be93861700dce517caf7ccf77358"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab5201096c45cc26ec3a47c7cd98473a4"><td class="memItemLeft" align="right" valign="top">string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ab5201096c45cc26ec3a47c7cd98473a4">GetFullErrorMessage</a> () const</td></tr>
<tr class="separator:ab5201096c45cc26ec3a47c7cd98473a4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a29ba3acd2c1e8ed61e8e0c367cc6051c"><td class="memItemLeft" align="right" valign="top">const string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a29ba3acd2c1e8ed61e8e0c367cc6051c">GetErrorDetails</a> () const</td></tr>
<tr class="memdesc:a29ba3acd2c1e8ed61e8e0c367cc6051c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns extended details about the last error. (e.g. error messages produced during file read operations that are otherwise lost.)  <a href="#a29ba3acd2c1e8ed61e8e0c367cc6051c">More...</a><br /></td></tr>
<tr class="separator:a29ba3acd2c1e8ed61e8e0c367cc6051c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae45785f80ecd7b6b7783fb290dd78152"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ae45785f80ecd7b6b7783fb290dd78152">SetErrorDetails</a> (const string &amp;details)</td></tr>
<tr class="memdesc:ae45785f80ecd7b6b7783fb290dd78152"><td class="mdescLeft">&#160;</td><td class="mdescRight">Set extended details about the last error. (e.g. error messages produced during file read operations that are otherwise lost.)  <a href="#ae45785f80ecd7b6b7783fb290dd78152">More...</a><br /></td></tr>
<tr class="separator:ae45785f80ecd7b6b7783fb290dd78152"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa746e33612c6cc7fcf8856fac3d364ac"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#aa746e33612c6cc7fcf8856fac3d364ac">SetSequenceLabel</a> (const string &amp;label)</td></tr>
<tr class="memdesc:aa746e33612c6cc7fcf8856fac3d364ac"><td class="mdescLeft">&#160;</td><td class="mdescRight">Set the name of the sequence. (This must be called before structure prediction for it to affect the resulting structures.)  <a href="#aa746e33612c6cc7fcf8856fac3d364ac">More...</a><br /></td></tr>
<tr class="separator:aa746e33612c6cc7fcf8856fac3d364ac"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3e1f322c241c729f36b56cb0e452c0c4"><td class="memItemLeft" align="right" valign="top">string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4">GetErrorMessageString</a> (const int error) const</td></tr>
<tr class="memdesc:a3e1f322c241c729f36b56cb0e452c0c4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return error messages based on code from GetErrorCode and other error codes.  <a href="#a3e1f322c241c729f36b56cb0e452c0c4">More...</a><br /></td></tr>
<tr class="separator:a3e1f322c241c729f36b56cb0e452c0c4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5e8cc3e54438c3a4ed2a042723386006"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a5e8cc3e54438c3a4ed2a042723386006">ResetError</a> ()</td></tr>
<tr class="separator:a5e8cc3e54438c3a4ed2a042723386006"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afdca3d7732fbffe48dfc7d52f193798b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#afdca3d7732fbffe48dfc7d52f193798b">EnsureStructureCapcacity</a> (const int minimumStructures)</td></tr>
<tr class="memdesc:afdca3d7732fbffe48dfc7d52f193798b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Ensure that at a minumum number of structures have been created.  <a href="#afdca3d7732fbffe48dfc7d52f193798b">More...</a><br /></td></tr>
<tr class="separator:afdca3d7732fbffe48dfc7d52f193798b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac529346af902b21f9b3bb2b17a3ee661"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ac529346af902b21f9b3bb2b17a3ee661">SpecifyPair</a> (const int i, const int j, const int structurenumber=1)</td></tr>
<tr class="memdesc:ac529346af902b21f9b3bb2b17a3ee661"><td class="mdescLeft">&#160;</td><td class="mdescRight">Specify a base pair between nucleotides i and j.  <a href="#ac529346af902b21f9b3bb2b17a3ee661">More...</a><br /></td></tr>
<tr class="separator:ac529346af902b21f9b3bb2b17a3ee661"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab6f64d9425687dc03f7cedf417bc9bc3"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ab6f64d9425687dc03f7cedf417bc9bc3">RemovePairs</a> (const int structurenumber=1, bool removeIfLastStructure=true)</td></tr>
<tr class="memdesc:ab6f64d9425687dc03f7cedf417bc9bc3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Remove all the current base pairs in a specified structure.  <a href="#ab6f64d9425687dc03f7cedf417bc9bc3">More...</a><br /></td></tr>
<tr class="separator:ab6f64d9425687dc03f7cedf417bc9bc3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7aae4a7dcb9fcd87d3cf0c4c2393b9b3"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a7aae4a7dcb9fcd87d3cf0c4c2393b9b3">RemoveBasePair</a> (const int i, const int structurenumber=1)</td></tr>
<tr class="memdesc:a7aae4a7dcb9fcd87d3cf0c4c2393b9b3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Remove a specified pair in a specified structure.  <a href="#a7aae4a7dcb9fcd87d3cf0c4c2393b9b3">More...</a><br /></td></tr>
<tr class="separator:a7aae4a7dcb9fcd87d3cf0c4c2393b9b3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a815bd1662622d7588a80107777e713be"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a815bd1662622d7588a80107777e713be">CalculateFreeEnergy</a> (const int structurenumber=1, const bool UseSimpleMBLoopRules=false)</td></tr>
<tr class="memdesc:a815bd1662622d7588a80107777e713be"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the predicted Gibb's free energy change for structure # structurenumber, defaulted to 1.  <a href="#a815bd1662622d7588a80107777e713be">More...</a><br /></td></tr>
<tr class="separator:a815bd1662622d7588a80107777e713be"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa61adf54c3d44cc6e29f6ae523342491"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#aa61adf54c3d44cc6e29f6ae523342491">ExteriorLoopCorrection</a> (const int structurenumber, const bool UseSimpleMBLoopRules, int min_index, int max_index)</td></tr>
<tr class="separator:aa61adf54c3d44cc6e29f6ae523342491"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab08f9c44b784e65ab966583e420fcba4"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ab08f9c44b784e65ab966583e420fcba4">WriteThermodynamicDetails</a> (const char filename[], const bool UseSimpleMBLoopRules=false)</td></tr>
<tr class="memdesc:ab08f9c44b784e65ab966583e420fcba4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate the folding free energy change for all structures and write the details of the calculation to a file.  <a href="#ab08f9c44b784e65ab966583e420fcba4">More...</a><br /></td></tr>
<tr class="separator:ab08f9c44b784e65ab966583e420fcba4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaa9563532ecfeee0de773396b8fc9e17"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#aaa9563532ecfeee0de773396b8fc9e17">FoldSingleStrand</a> (const float percent=20, const int maximumstructures=20, const int window=5, const char savefile[]=&quot;&quot;, const int maxinternalloopsize=30, bool mfeonly=false, bool simple_iloops=true, bool disablecoax=false, bool allow_isolated=false)</td></tr>
<tr class="memdesc:aaa9563532ecfeee0de773396b8fc9e17"><td class="mdescLeft">&#160;</td><td class="mdescRight">Predict the lowest free energy secondary structure and generate suboptimal structures using a heuristic.  <a href="#aaa9563532ecfeee0de773396b8fc9e17">More...</a><br /></td></tr>
<tr class="separator:aaa9563532ecfeee0de773396b8fc9e17"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae9c16cf6e875e5152eca3f4b94721fe3"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ae9c16cf6e875e5152eca3f4b94721fe3">GenerateAllSuboptimalStructures</a> (const float percent=5, const double deltaG=0.6)</td></tr>
<tr class="memdesc:ae9c16cf6e875e5152eca3f4b94721fe3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Predict the lowest free energy secondary structure and generate all suboptimal structures.  <a href="#ae9c16cf6e875e5152eca3f4b94721fe3">More...</a><br /></td></tr>
<tr class="separator:ae9c16cf6e875e5152eca3f4b94721fe3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a61141a694126843efe07081e5918fba7"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a61141a694126843efe07081e5918fba7">MaximizeExpectedAccuracy</a> (const double maxPercent=20, const int maxStructures=20, const int window=1, const double gamma=1.0)</td></tr>
<tr class="memdesc:a61141a694126843efe07081e5918fba7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Predict the structure with maximum expected accuracy and suboptimal structures.  <a href="#a61141a694126843efe07081e5918fba7">More...</a><br /></td></tr>
<tr class="separator:a61141a694126843efe07081e5918fba7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a654db191fe8a0a76350dfc57b2c5740c"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a654db191fe8a0a76350dfc57b2c5740c">PartitionFunction</a> (const char savefile[]=&quot;&quot;, double temperature=-10.0, bool disablecoax=false, bool restoreSHAPE=true, bool allowisolated=false, int maxinter=30)</td></tr>
<tr class="memdesc:a654db191fe8a0a76350dfc57b2c5740c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Predict the partition function for a sequence.  <a href="#a654db191fe8a0a76350dfc57b2c5740c">More...</a><br /></td></tr>
<tr class="separator:a654db191fe8a0a76350dfc57b2c5740c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aac5de92b17d1fe4d2eb0cc3fbee595aa"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#aac5de92b17d1fe4d2eb0cc3fbee595aa">Rsample</a> (const vector&lt; double &gt; &amp;experimentalRestraints, RsampleData &amp;refdata, const int randomSeed=0, const char savefile[]=&quot;&quot;, const double cparam=0.5, const double offset=1.10, const int numsamples=10000)</td></tr>
<tr class="separator:aac5de92b17d1fe4d2eb0cc3fbee595aa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad05388cc60648f09cc1acb3c4b513516"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ad05388cc60648f09cc1acb3c4b513516">PredictProbablePairs</a> (const float probability=0)</td></tr>
<tr class="memdesc:ad05388cc60648f09cc1acb3c4b513516"><td class="mdescLeft">&#160;</td><td class="mdescRight">Predict structures containing highly probable pairs.  <a href="#ad05388cc60648f09cc1acb3c4b513516">More...</a><br /></td></tr>
<tr class="separator:ad05388cc60648f09cc1acb3c4b513516"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab5dd63132dc8324c84c670cf887884c2"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ab5dd63132dc8324c84c670cf887884c2">ProbKnot</a> (int iterations=1, int MinHelixLength=1, double threshold=0)</td></tr>
<tr class="memdesc:ab5dd63132dc8324c84c670cf887884c2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Predict maximum expected accuracy structures that contain pseudoknots from either a sequence or a partition function save file.  <a href="#ab5dd63132dc8324c84c670cf887884c2">More...</a><br /></td></tr>
<tr class="separator:ab5dd63132dc8324c84c670cf887884c2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2049ad01ea76e5072133e23dbd24d03d"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a2049ad01ea76e5072133e23dbd24d03d">ProbKnotFromSample</a> (int iterations=1, int MinHelixLength=1, double threshold=0)</td></tr>
<tr class="memdesc:a2049ad01ea76e5072133e23dbd24d03d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Predict maximum expected accuracy structures that contain pseudoknots from a file containing ensemble of structures.  <a href="#a2049ad01ea76e5072133e23dbd24d03d">More...</a><br /></td></tr>
<tr class="separator:a2049ad01ea76e5072133e23dbd24d03d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac068513a05333629af4fad23a5b3213d"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ac068513a05333629af4fad23a5b3213d">ReFoldSingleStrand</a> (const float percent=20, const int maximumstructures=20, const int window=5)</td></tr>
<tr class="memdesc:ac068513a05333629af4fad23a5b3213d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Re-predict the lowest free energy secondary structure and generate suboptimal structures using a heuristic.  <a href="#ac068513a05333629af4fad23a5b3213d">More...</a><br /></td></tr>
<tr class="separator:ac068513a05333629af4fad23a5b3213d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab06ab0d0af7f74d3e3a20c8d52e65ef5"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ab06ab0d0af7f74d3e3a20c8d52e65ef5">Stochastic</a> (const int structures=1000, const int seed=1)</td></tr>
<tr class="memdesc:ab06ab0d0af7f74d3e3a20c8d52e65ef5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Sample structures from the Boltzman ensemable.  <a href="#ab06ab0d0af7f74d3e3a20c8d52e65ef5">More...</a><br /></td></tr>
<tr class="separator:ab06ab0d0af7f74d3e3a20c8d52e65ef5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8ba352350bcda5514d772db7e763229e"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a8ba352350bcda5514d772db7e763229e">ForceDoubleStranded</a> (const int i)</td></tr>
<tr class="memdesc:a8ba352350bcda5514d772db7e763229e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Force a nucleotide to be double stranded (base paired).  <a href="#a8ba352350bcda5514d772db7e763229e">More...</a><br /></td></tr>
<tr class="separator:a8ba352350bcda5514d772db7e763229e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae2ecc054cf728366d884ecf8f066ee20"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ae2ecc054cf728366d884ecf8f066ee20">ForceFMNCleavage</a> (const int i)</td></tr>
<tr class="memdesc:ae2ecc054cf728366d884ecf8f066ee20"><td class="mdescLeft">&#160;</td><td class="mdescRight">Indicate a nucleotide that is accessible to FMN cleavage (a U in GU pair).  <a href="#ae2ecc054cf728366d884ecf8f066ee20">More...</a><br /></td></tr>
<tr class="separator:ae2ecc054cf728366d884ecf8f066ee20"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a99f37cbc69b7ee0e99116c6fe074e7a7"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a99f37cbc69b7ee0e99116c6fe074e7a7">ForceMaximumPairingDistance</a> (const int distance)</td></tr>
<tr class="memdesc:a99f37cbc69b7ee0e99116c6fe074e7a7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Force a maximum distance between apired nucleotides.  <a href="#a99f37cbc69b7ee0e99116c6fe074e7a7">More...</a><br /></td></tr>
<tr class="separator:a99f37cbc69b7ee0e99116c6fe074e7a7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac6f531c36afbcbb44dfad283a31c5879"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ac6f531c36afbcbb44dfad283a31c5879">ForceModification</a> (const int i)</td></tr>
<tr class="memdesc:ac6f531c36afbcbb44dfad283a31c5879"><td class="mdescLeft">&#160;</td><td class="mdescRight">Force modification for a nucleotide.  <a href="#ac6f531c36afbcbb44dfad283a31c5879">More...</a><br /></td></tr>
<tr class="separator:ac6f531c36afbcbb44dfad283a31c5879"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6d33a787f66bbef7e78cf64624c0e12e"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a6d33a787f66bbef7e78cf64624c0e12e">ForcePair</a> (const int i, const int j)</td></tr>
<tr class="memdesc:a6d33a787f66bbef7e78cf64624c0e12e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Force a pair between two nucleotides.  <a href="#a6d33a787f66bbef7e78cf64624c0e12e">More...</a><br /></td></tr>
<tr class="separator:a6d33a787f66bbef7e78cf64624c0e12e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4fa207c587f788981d860f70616055e4"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a4fa207c587f788981d860f70616055e4">ForceProhibitPair</a> (const int i, const int j)</td></tr>
<tr class="memdesc:a4fa207c587f788981d860f70616055e4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prohibit a pair between two nucleotides.  <a href="#a4fa207c587f788981d860f70616055e4">More...</a><br /></td></tr>
<tr class="separator:a4fa207c587f788981d860f70616055e4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aef005a05b220b79044a620d202aa8d1b"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#aef005a05b220b79044a620d202aa8d1b">ForceSingleStranded</a> (const int i)</td></tr>
<tr class="memdesc:aef005a05b220b79044a620d202aa8d1b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Force a nucleotide to be single stranded.  <a href="#aef005a05b220b79044a620d202aa8d1b">More...</a><br /></td></tr>
<tr class="separator:aef005a05b220b79044a620d202aa8d1b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7e1d3fc6f19b4f5c8fe5373763996029"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a7e1d3fc6f19b4f5c8fe5373763996029">GetForcedDoubleStranded</a> (const int constraintnumber)</td></tr>
<tr class="memdesc:a7e1d3fc6f19b4f5c8fe5373763996029"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return a nucleotide that is forced double stranded.  <a href="#a7e1d3fc6f19b4f5c8fe5373763996029">More...</a><br /></td></tr>
<tr class="separator:a7e1d3fc6f19b4f5c8fe5373763996029"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a85711931f3d25f6c2f209ec64beba01c"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a85711931f3d25f6c2f209ec64beba01c">GetForcedFMNCleavage</a> (const int constraintnumber)</td></tr>
<tr class="memdesc:a85711931f3d25f6c2f209ec64beba01c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return a nucleotide that is accessible to FMN cleavage.  <a href="#a85711931f3d25f6c2f209ec64beba01c">More...</a><br /></td></tr>
<tr class="separator:a85711931f3d25f6c2f209ec64beba01c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9c1eb7ad1992c12d69979418a14ba883"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a9c1eb7ad1992c12d69979418a14ba883">GetForcedModification</a> (const int constraintnumber)</td></tr>
<tr class="memdesc:a9c1eb7ad1992c12d69979418a14ba883"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return a nucleotide that is accessible to modification.  <a href="#a9c1eb7ad1992c12d69979418a14ba883">More...</a><br /></td></tr>
<tr class="separator:a9c1eb7ad1992c12d69979418a14ba883"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1d459e60cbbd1d9648a3b61189c76c8d"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a1d459e60cbbd1d9648a3b61189c76c8d">GetForcedPair</a> (const int constraintnumber, const bool fiveprime)</td></tr>
<tr class="memdesc:a1d459e60cbbd1d9648a3b61189c76c8d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return a nucleotide in a forced pair.  <a href="#a1d459e60cbbd1d9648a3b61189c76c8d">More...</a><br /></td></tr>
<tr class="separator:a1d459e60cbbd1d9648a3b61189c76c8d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad993910aa7abba8e80cf3cbb590bddab"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ad993910aa7abba8e80cf3cbb590bddab">GetForcedProhibitedPair</a> (const int constraintnumber, const bool fiveprime)</td></tr>
<tr class="memdesc:ad993910aa7abba8e80cf3cbb590bddab"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return a nucleotide in a prohibited pair.  <a href="#ad993910aa7abba8e80cf3cbb590bddab">More...</a><br /></td></tr>
<tr class="separator:ad993910aa7abba8e80cf3cbb590bddab"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abe8372147747e249430353f7366c4310"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#abe8372147747e249430353f7366c4310">GetForcedSingleStranded</a> (const int constraintnumber)</td></tr>
<tr class="memdesc:abe8372147747e249430353f7366c4310"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return a nucleotide that is forced single stranded.  <a href="#abe8372147747e249430353f7366c4310">More...</a><br /></td></tr>
<tr class="separator:abe8372147747e249430353f7366c4310"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3bc6d5801335f85eed40a2daa987d96b"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a3bc6d5801335f85eed40a2daa987d96b">GetMaximumPairingDistance</a> ()</td></tr>
<tr class="memdesc:a3bc6d5801335f85eed40a2daa987d96b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the maximum pairing distance.  <a href="#a3bc6d5801335f85eed40a2daa987d96b">More...</a><br /></td></tr>
<tr class="separator:a3bc6d5801335f85eed40a2daa987d96b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af9730b22a5500a3f23b4ee80382074e5"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#af9730b22a5500a3f23b4ee80382074e5">GetNumberOfForcedDoubleStranded</a> ()</td></tr>
<tr class="memdesc:af9730b22a5500a3f23b4ee80382074e5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the number of nucletides forced to be paired.  <a href="#af9730b22a5500a3f23b4ee80382074e5">More...</a><br /></td></tr>
<tr class="separator:af9730b22a5500a3f23b4ee80382074e5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8a3843c206b1cf6790754454da102da4"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a8a3843c206b1cf6790754454da102da4">GetNumberOfForcedFMNCleavages</a> ()</td></tr>
<tr class="memdesc:a8a3843c206b1cf6790754454da102da4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the number of nucleotides accessible to FMN cleavage.  <a href="#a8a3843c206b1cf6790754454da102da4">More...</a><br /></td></tr>
<tr class="separator:a8a3843c206b1cf6790754454da102da4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a50c7d53681d5618f3c04e2140f8bb0c5"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a50c7d53681d5618f3c04e2140f8bb0c5">GetNumberOfForcedModifications</a> ()</td></tr>
<tr class="memdesc:a50c7d53681d5618f3c04e2140f8bb0c5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the number of nucleotides accessible to chemical modification.  <a href="#a50c7d53681d5618f3c04e2140f8bb0c5">More...</a><br /></td></tr>
<tr class="separator:a50c7d53681d5618f3c04e2140f8bb0c5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae1c0ee1877897c83292d4194cc137206"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ae1c0ee1877897c83292d4194cc137206">GetNumberOfForcedPairs</a> ()</td></tr>
<tr class="memdesc:ae1c0ee1877897c83292d4194cc137206"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the number of forced base pairs.  <a href="#ae1c0ee1877897c83292d4194cc137206">More...</a><br /></td></tr>
<tr class="separator:ae1c0ee1877897c83292d4194cc137206"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a11cd24997b6494ed631121b84e3b9b9b"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a11cd24997b6494ed631121b84e3b9b9b">GetNumberOfForcedProhibitedPairs</a> ()</td></tr>
<tr class="memdesc:a11cd24997b6494ed631121b84e3b9b9b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the number of prohibited base pairs.  <a href="#a11cd24997b6494ed631121b84e3b9b9b">More...</a><br /></td></tr>
<tr class="separator:a11cd24997b6494ed631121b84e3b9b9b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a505c57ee97ac3b5c1c21063087658ca3"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a505c57ee97ac3b5c1c21063087658ca3">GetNumberOfForcedSingleStranded</a> ()</td></tr>
<tr class="memdesc:a505c57ee97ac3b5c1c21063087658ca3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the number of nucleotides that are not allowed to pair.  <a href="#a505c57ee97ac3b5c1c21063087658ca3">More...</a><br /></td></tr>
<tr class="separator:a505c57ee97ac3b5c1c21063087658ca3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a256ed0d8b5a7fdd1dc1ea5ced058cc7a"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a256ed0d8b5a7fdd1dc1ea5ced058cc7a">ReadConstraints</a> (const char filename[])</td></tr>
<tr class="memdesc:a256ed0d8b5a7fdd1dc1ea5ced058cc7a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Read a set of folding constraints to disk in a plain text file.  <a href="#a256ed0d8b5a7fdd1dc1ea5ced058cc7a">More...</a><br /></td></tr>
<tr class="separator:a256ed0d8b5a7fdd1dc1ea5ced058cc7a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a54e154d693ade8fcfe62e47961d37701"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a54e154d693ade8fcfe62e47961d37701">ReadSHAPE</a> (const char filename[], const double slope, const double intercept, RestraintType modifier=RESTRAINT_SHAPE, const bool IsPseudoEnergy=true)</td></tr>
<tr class="memdesc:a54e154d693ade8fcfe62e47961d37701"><td class="mdescLeft">&#160;</td><td class="mdescRight">Read SHAPE data from disk.  <a href="#a54e154d693ade8fcfe62e47961d37701">More...</a><br /></td></tr>
<tr class="separator:a54e154d693ade8fcfe62e47961d37701"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adb84ef498e9d26d3f7add0e47855621a"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#adb84ef498e9d26d3f7add0e47855621a">ReadSHAPE</a> (const char filename[], const double dsSlope, const double dsIntercept, const double ssSlope, const double ssIntercept, RestraintType modifier=RESTRAINT_SHAPE)</td></tr>
<tr class="memdesc:adb84ef498e9d26d3f7add0e47855621a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Read SHAPE data from disk including single-stranded SHAPE pseudo free energys.  <a href="#adb84ef498e9d26d3f7add0e47855621a">More...</a><br /></td></tr>
<tr class="separator:adb84ef498e9d26d3f7add0e47855621a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0a8bba2d4d6c3a9afd718a7bf420eb29"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a0a8bba2d4d6c3a9afd718a7bf420eb29">ReadDMS</a> (const char filename[], const bool bynt=false)</td></tr>
<tr class="separator:a0a8bba2d4d6c3a9afd718a7bf420eb29"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a05b4ec1bf9e25c5673c065baabab6233"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a05b4ec1bf9e25c5673c065baabab6233">ReadDSO</a> (const char filename[])</td></tr>
<tr class="memdesc:a05b4ec1bf9e25c5673c065baabab6233"><td class="mdescLeft">&#160;</td><td class="mdescRight">Read double strand offset data from disk.  <a href="#a05b4ec1bf9e25c5673c065baabab6233">More...</a><br /></td></tr>
<tr class="separator:a05b4ec1bf9e25c5673c065baabab6233"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6ed8d145eb78bfcb22aa5f5e7e41f7c8"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a6ed8d145eb78bfcb22aa5f5e7e41f7c8">ReadSSO</a> (const char filename[])</td></tr>
<tr class="memdesc:a6ed8d145eb78bfcb22aa5f5e7e41f7c8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Read single strand offset data from disk.  <a href="#a6ed8d145eb78bfcb22aa5f5e7e41f7c8">More...</a><br /></td></tr>
<tr class="separator:a6ed8d145eb78bfcb22aa5f5e7e41f7c8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad15624f880c955d1985914292b95d746"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ad15624f880c955d1985914292b95d746">ReadExperimentalPairBonus</a> (const char filename[], double const experimentalOffset, double const experimentalScaling)</td></tr>
<tr class="memdesc:ad15624f880c955d1985914292b95d746"><td class="mdescLeft">&#160;</td><td class="mdescRight">Read experimental pair bonuses from disk.  <a href="#ad15624f880c955d1985914292b95d746">More...</a><br /></td></tr>
<tr class="separator:ad15624f880c955d1985914292b95d746"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3b4a3b4f525013b59f48b63423108d3c"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a3b4a3b4f525013b59f48b63423108d3c">RemoveConstraints</a> ()</td></tr>
<tr class="memdesc:a3b4a3b4f525013b59f48b63423108d3c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Remove all folding constraints.  <a href="#a3b4a3b4f525013b59f48b63423108d3c">More...</a><br /></td></tr>
<tr class="separator:a3b4a3b4f525013b59f48b63423108d3c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac7a793cd3862b976a5b354da43a60df2"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ac7a793cd3862b976a5b354da43a60df2">SetConstraints</a> (vector&lt; int &gt; ss)</td></tr>
<tr class="separator:ac7a793cd3862b976a5b354da43a60df2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a127218155fa5e2d68aaf6729e97616e0"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a127218155fa5e2d68aaf6729e97616e0">SetExtrinsic</a> (int i, int j, double k)</td></tr>
<tr class="memdesc:a127218155fa5e2d68aaf6729e97616e0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Add extrinsic restraints for partition function calculations.  <a href="#a127218155fa5e2d68aaf6729e97616e0">More...</a><br /></td></tr>
<tr class="separator:a127218155fa5e2d68aaf6729e97616e0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa860f3f448840a8af8461212b1920ab6"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#aa860f3f448840a8af8461212b1920ab6">WriteConstraints</a> (const char filename[])</td></tr>
<tr class="memdesc:aa860f3f448840a8af8461212b1920ab6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Write the current set of folding constraints to disk in a plain text file.  <a href="#aa860f3f448840a8af8461212b1920ab6">More...</a><br /></td></tr>
<tr class="separator:aa860f3f448840a8af8461212b1920ab6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1eebbd7916abda62860c611e04bd977d"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a1eebbd7916abda62860c611e04bd977d">AddComment</a> (const char comment[], const int structurenumber=1)</td></tr>
<tr class="memdesc:a1eebbd7916abda62860c611e04bd977d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Add a comment associated with a structure.  <a href="#a1eebbd7916abda62860c611e04bd977d">More...</a><br /></td></tr>
<tr class="separator:a1eebbd7916abda62860c611e04bd977d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6e1e0db7f8f0f3d9cb7dc14f5dc61953"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a6e1e0db7f8f0f3d9cb7dc14f5dc61953">WriteCt</a> (const char filename[], bool append=false, CTCommentProvider &amp;commentProvider=CTComments::Energy) const</td></tr>
<tr class="memdesc:a6e1e0db7f8f0f3d9cb7dc14f5dc61953"><td class="mdescLeft">&#160;</td><td class="mdescRight">Write a ct file of the structures.  <a href="#a6e1e0db7f8f0f3d9cb7dc14f5dc61953">More...</a><br /></td></tr>
<tr class="separator:a6e1e0db7f8f0f3d9cb7dc14f5dc61953"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2c04e45d3cf6436afe8622d140f2f1d1"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a2c04e45d3cf6436afe8622d140f2f1d1">WriteDotBracket</a> (const char filename[], const int structurenumber=-1, const DotBracketFormat format=DBN_FMT_MULTI_TITLE, CTCommentProvider &amp;commentProvider=CTComments::Energy) const</td></tr>
<tr class="memdesc:a2c04e45d3cf6436afe8622d140f2f1d1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Write dot-bracket file of structures.  <a href="#a2c04e45d3cf6436afe8622d140f2f1d1">More...</a><br /></td></tr>
<tr class="separator:a2c04e45d3cf6436afe8622d140f2f1d1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae052a7062e9eb11c74027eba2f41d5d7"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ae052a7062e9eb11c74027eba2f41d5d7">BreakPseudoknot</a> (const bool minimum_energy=true, const int structurenumber=0, const bool useFastMethod=true)</td></tr>
<tr class="memdesc:ae052a7062e9eb11c74027eba2f41d5d7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Break any pseudoknots that might be in a structure.  <a href="#ae052a7062e9eb11c74027eba2f41d5d7">More...</a><br /></td></tr>
<tr class="separator:ae052a7062e9eb11c74027eba2f41d5d7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abee55f46e52b836f32c9a572af8e2d7c"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#abee55f46e52b836f32c9a572af8e2d7c">ContainsPseudoknot</a> (const int structurenumber)</td></tr>
<tr class="memdesc:abee55f46e52b836f32c9a572af8e2d7c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Report if there are any pseudoknots in a structure.  <a href="#abee55f46e52b836f32c9a572af8e2d7c">More...</a><br /></td></tr>
<tr class="separator:abee55f46e52b836f32c9a572af8e2d7c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a88e712e84b77792e813d22b7a51779f1"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a88e712e84b77792e813d22b7a51779f1">GetEnsembleEnergy</a> ()</td></tr>
<tr class="memdesc:a88e712e84b77792e813d22b7a51779f1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the ensemble folding free energy change.  <a href="#a88e712e84b77792e813d22b7a51779f1">More...</a><br /></td></tr>
<tr class="separator:a88e712e84b77792e813d22b7a51779f1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a08789d87f4a205b16f643696206c5eb4"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a08789d87f4a205b16f643696206c5eb4">GetEnsembleDefect</a> (const int structurenumber=1, const int start=0, const int end=0, const char filename[]=&quot;&quot;)</td></tr>
<tr class="memdesc:a08789d87f4a205b16f643696206c5eb4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the ensemble defect of a secondary structure.  <a href="#a08789d87f4a205b16f643696206c5eb4">More...</a><br /></td></tr>
<tr class="separator:a08789d87f4a205b16f643696206c5eb4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4534199d60cfffe08f0a89cd77f3acd9"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a4534199d60cfffe08f0a89cd77f3acd9">GetFreeEnergy</a> (const int structurenumber)</td></tr>
<tr class="memdesc:a4534199d60cfffe08f0a89cd77f3acd9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the folding free energy change for a predicted structure.  <a href="#a4534199d60cfffe08f0a89cd77f3acd9">More...</a><br /></td></tr>
<tr class="separator:a4534199d60cfffe08f0a89cd77f3acd9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7dfa253c1491b51383acd12b73c72a2f"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a7dfa253c1491b51383acd12b73c72a2f">GetPair</a> (const int i, const int structurenumber=1)</td></tr>
<tr class="memdesc:a7dfa253c1491b51383acd12b73c72a2f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the nucleotide to which the specified nucleotide is paired.  <a href="#a7dfa253c1491b51383acd12b73c72a2f">More...</a><br /></td></tr>
<tr class="separator:a7dfa253c1491b51383acd12b73c72a2f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa240d53bc9bc53897931a18b67bbcee0"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#aa240d53bc9bc53897931a18b67bbcee0">GetPairEnergy</a> (const int i, const int j)</td></tr>
<tr class="memdesc:aa240d53bc9bc53897931a18b67bbcee0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the lowest folding free energy possible for a structure containing pair i-j.  <a href="#aa240d53bc9bc53897931a18b67bbcee0">More...</a><br /></td></tr>
<tr class="separator:aa240d53bc9bc53897931a18b67bbcee0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abd1da0a82663b38414a01c39b1a3973e"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#abd1da0a82663b38414a01c39b1a3973e">GetPairProbability</a> (const int i, const int j)</td></tr>
<tr class="memdesc:abd1da0a82663b38414a01c39b1a3973e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get a base pair probability.  <a href="#abd1da0a82663b38414a01c39b1a3973e">More...</a><br /></td></tr>
<tr class="separator:abd1da0a82663b38414a01c39b1a3973e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a07e1fb189a45eb5b4604c286d5dd9d14"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a07e1fb189a45eb5b4604c286d5dd9d14">GetPairProbabilities</a> (double *arr, const int size)</td></tr>
<tr class="separator:a07e1fb189a45eb5b4604c286d5dd9d14"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a27b3bdb17fb4233cc4f91753beda7570"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a27b3bdb17fb4233cc4f91753beda7570">GetStructureNumber</a> () const</td></tr>
<tr class="memdesc:a27b3bdb17fb4233cc4f91753beda7570"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the total number of specified or predicted structures.  <a href="#a27b3bdb17fb4233cc4f91753beda7570">More...</a><br /></td></tr>
<tr class="separator:a27b3bdb17fb4233cc4f91753beda7570"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae464f8408d0db524408ab27cd6ed5f6e"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ae464f8408d0db524408ab27cd6ed5f6e">DetermineDrawingCoordinates</a> (const int height, const int width, const int structurenumber=1)</td></tr>
<tr class="memdesc:ae464f8408d0db524408ab27cd6ed5f6e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Determine the coordinates for drawing a secondary structure.  <a href="#ae464f8408d0db524408ab27cd6ed5f6e">More...</a><br /></td></tr>
<tr class="separator:ae464f8408d0db524408ab27cd6ed5f6e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab329d510a2ce6765348142c8a6f7dcaa"><td class="memItemLeft" align="right" valign="top">std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ab329d510a2ce6765348142c8a6f7dcaa">GetCommentString</a> (const int structurenumber=1)</td></tr>
<tr class="memdesc:ab329d510a2ce6765348142c8a6f7dcaa"><td class="mdescLeft">&#160;</td><td class="mdescRight">Provide the comment from the ct file as a string.  <a href="#ab329d510a2ce6765348142c8a6f7dcaa">More...</a><br /></td></tr>
<tr class="separator:ab329d510a2ce6765348142c8a6f7dcaa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a97c3e9a51d32a73fadd53f467c8f3ae0"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a97c3e9a51d32a73fadd53f467c8f3ae0">GetNucleotideXCoordinate</a> (const int i)</td></tr>
<tr class="memdesc:a97c3e9a51d32a73fadd53f467c8f3ae0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the X coordinate for nucleotide i for drawing a structure.  <a href="#a97c3e9a51d32a73fadd53f467c8f3ae0">More...</a><br /></td></tr>
<tr class="separator:a97c3e9a51d32a73fadd53f467c8f3ae0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acb2539f11b34a8e34668697ebf2890fb"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#acb2539f11b34a8e34668697ebf2890fb">GetNucleotideYCoordinate</a> (const int i)</td></tr>
<tr class="memdesc:acb2539f11b34a8e34668697ebf2890fb"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the Y coordinate for nucleotide i for drawing a structure.  <a href="#acb2539f11b34a8e34668697ebf2890fb">More...</a><br /></td></tr>
<tr class="separator:acb2539f11b34a8e34668697ebf2890fb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a23c879c487fcb254ebf645096026802e"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a23c879c487fcb254ebf645096026802e">GetLabelXCoordinate</a> (const int i)</td></tr>
<tr class="memdesc:a23c879c487fcb254ebf645096026802e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the X coordinate for placing the nucleotide index label specified by i.  <a href="#a23c879c487fcb254ebf645096026802e">More...</a><br /></td></tr>
<tr class="separator:a23c879c487fcb254ebf645096026802e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:affd3ef139bea53fbc39df41f68e1f030"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#affd3ef139bea53fbc39df41f68e1f030">GetLabelYCoordinate</a> (const int i)</td></tr>
<tr class="memdesc:affd3ef139bea53fbc39df41f68e1f030"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the Y coordinate for placing the nucleotide index label specified by i.  <a href="#affd3ef139bea53fbc39df41f68e1f030">More...</a><br /></td></tr>
<tr class="separator:affd3ef139bea53fbc39df41f68e1f030"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a84842b95ea62a357b12c440ad8ff4912"><td class="memItemLeft" align="right" valign="top">char&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a84842b95ea62a357b12c440ad8ff4912">GetNucleotide</a> (const int i)</td></tr>
<tr class="separator:a84842b95ea62a357b12c440ad8ff4912"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4ca21eb3e90dc475cc32c616b614ebc2"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a4ca21eb3e90dc475cc32c616b614ebc2">GetSequenceLength</a> () const</td></tr>
<tr class="memdesc:a4ca21eb3e90dc475cc32c616b614ebc2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the total length of the sequence.  <a href="#a4ca21eb3e90dc475cc32c616b614ebc2">More...</a><br /></td></tr>
<tr class="separator:a4ca21eb3e90dc475cc32c616b614ebc2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac3b16f4e3c72091bf69a55e923c51728"><td class="memItemLeft" align="right" valign="top">const char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ac3b16f4e3c72091bf69a55e923c51728">GetSequence</a> () const</td></tr>
<tr class="separator:ac3b16f4e3c72091bf69a55e923c51728"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9f49a0851278aaa5aa39a1d538c58fa9"><td class="memItemLeft" align="right" valign="top">std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a9f49a0851278aaa5aa39a1d538c58fa9">GetSequence</a> (size_t start, size_t length=std::string::npos) const</td></tr>
<tr class="separator:a9f49a0851278aaa5aa39a1d538c58fa9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a571a75afbe286a928b1910b19603273d"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a571a75afbe286a928b1910b19603273d">GetBackboneType</a> () const</td></tr>
<tr class="memdesc:a571a75afbe286a928b1910b19603273d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the backbone type.  <a href="#a571a75afbe286a928b1910b19603273d">More...</a><br /></td></tr>
<tr class="separator:a571a75afbe286a928b1910b19603273d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a78bdc62dac46d0d817fed67f5843010c"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a78bdc62dac46d0d817fed67f5843010c">GetVprimeQ</a> (const int i, const int j)</td></tr>
<tr class="separator:a78bdc62dac46d0d817fed67f5843010c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad1d51357d72c2be6b5f62c31594d401b"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ad1d51357d72c2be6b5f62c31594d401b">GetW</a> (const int i, const int j)</td></tr>
<tr class="separator:ad1d51357d72c2be6b5f62c31594d401b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7513721374fab949fd42bbe4980894dc"><td class="memItemLeft" align="right" valign="top">structure *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a7513721374fab949fd42bbe4980894dc">GetStructure</a> ()</td></tr>
<tr class="separator:a7513721374fab949fd42bbe4980894dc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a507cd74f3b8a98e2172ec7c85465d919"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a507cd74f3b8a98e2172ec7c85465d919">SetProgress</a> (ProgressHandler &amp;Progress)</td></tr>
<tr class="separator:a507cd74f3b8a98e2172ec7c85465d919"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae59b36ca61c62813ad096eebb02f3850"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ae59b36ca61c62813ad096eebb02f3850">StopProgress</a> ()</td></tr>
<tr class="separator:ae59b36ca61c62813ad096eebb02f3850"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a772f52c2865590b7a153ae3a6aab1de5"><td class="memItemLeft" align="right" valign="top">ProgressHandler *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a772f52c2865590b7a153ae3a6aab1de5">GetProgress</a> ()</td></tr>
<tr class="separator:a772f52c2865590b7a153ae3a6aab1de5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adf930d0e1dd46d4d7dedf006228e2bbf"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#adf930d0e1dd46d4d7dedf006228e2bbf">~RNA</a> ()</td></tr>
<tr class="memdesc:adf930d0e1dd46d4d7dedf006228e2bbf"><td class="mdescLeft">&#160;</td><td class="mdescRight">Destructor.  <a href="#adf930d0e1dd46d4d7dedf006228e2bbf">More...</a><br /></td></tr>
<tr class="separator:adf930d0e1dd46d4d7dedf006228e2bbf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a155d2c6ae7d094206ec0699228f67c46"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a155d2c6ae7d094206ec0699228f67c46">CopyThermo</a> (<a class="el" href="class_thermodynamics.html">Thermodynamics</a> &amp;copy)</td></tr>
<tr class="memdesc:a155d2c6ae7d094206ec0699228f67c46"><td class="mdescLeft">&#160;</td><td class="mdescRight">Copy thermodynamic parameters from an instance of an <a class="el" href="class_r_n_a.html">RNA</a> class.  <a href="#a155d2c6ae7d094206ec0699228f67c46">More...</a><br /></td></tr>
<tr class="separator:a155d2c6ae7d094206ec0699228f67c46"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="inherit_header pub_methods_class_thermodynamics"><td colspan="2" onclick="javascript:toggleInherit('pub_methods_class_thermodynamics')"><img src="closed.png" alt="-"/>&#160;Public Member Functions inherited from <a class="el" href="class_thermodynamics.html">Thermodynamics</a></td></tr>
<tr class="memitem:a83e22b26044cc4a49daaf67e05d9e506 inherit pub_methods_class_thermodynamics"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_thermodynamics.html#a83e22b26044cc4a49daaf67e05d9e506">Thermodynamics</a> (const bool isRNA=true, const char *const alphabetName=NULL, const double temperature=310.15)</td></tr>
<tr class="separator:a83e22b26044cc4a49daaf67e05d9e506 inherit pub_methods_class_thermodynamics"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a66af5db34dcd5f03cdb96e35ff42f40b inherit pub_methods_class_thermodynamics"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_thermodynamics.html#a66af5db34dcd5f03cdb96e35ff42f40b">Thermodynamics</a> (const <a class="el" href="class_thermodynamics.html">Thermodynamics</a> &amp;copyThermo)</td></tr>
<tr class="separator:a66af5db34dcd5f03cdb96e35ff42f40b inherit pub_methods_class_thermodynamics"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ada5e34040fbc505a85d3c5a833ca477d inherit pub_methods_class_thermodynamics"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_thermodynamics.html#ada5e34040fbc505a85d3c5a833ca477d">SetTemperature</a> (double temperature)</td></tr>
<tr class="memdesc:ada5e34040fbc505a85d3c5a833ca477d inherit pub_methods_class_thermodynamics"><td class="mdescLeft">&#160;</td><td class="mdescRight">Set the temperature of folding in K.  <a href="class_thermodynamics.html#ada5e34040fbc505a85d3c5a833ca477d">More...</a><br /></td></tr>
<tr class="separator:ada5e34040fbc505a85d3c5a833ca477d inherit pub_methods_class_thermodynamics"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8f5d357a80ae1742c70e9057c89a829e inherit pub_methods_class_thermodynamics"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_thermodynamics.html#a8f5d357a80ae1742c70e9057c89a829e">GetTemperature</a> () const</td></tr>
<tr class="separator:a8f5d357a80ae1742c70e9057c89a829e inherit pub_methods_class_thermodynamics"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af8b0c00a2b1d58663bd41eefa31fab3e inherit pub_methods_class_thermodynamics"><td class="memItemLeft" align="right" valign="top">string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_thermodynamics.html#af8b0c00a2b1d58663bd41eefa31fab3e">GetAlphabetName</a> () const</td></tr>
<tr class="memdesc:af8b0c00a2b1d58663bd41eefa31fab3e inherit pub_methods_class_thermodynamics"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the name of the extended alphabet for which thermodynamic parameters should be loaded.  <a href="class_thermodynamics.html#af8b0c00a2b1d58663bd41eefa31fab3e">More...</a><br /></td></tr>
<tr class="separator:af8b0c00a2b1d58663bd41eefa31fab3e inherit pub_methods_class_thermodynamics"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7aedcc0352524f85eeabd6ea89b299c8 inherit pub_methods_class_thermodynamics"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_thermodynamics.html#a7aedcc0352524f85eeabd6ea89b299c8">ReadThermodynamic</a> (const char *directory=NULL, const char *alphabet=NULL, const double temperature=-1.0)</td></tr>
<tr class="memdesc:a7aedcc0352524f85eeabd6ea89b299c8 inherit pub_methods_class_thermodynamics"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to read the thermodynamic parameters.  <a href="class_thermodynamics.html#a7aedcc0352524f85eeabd6ea89b299c8">More...</a><br /></td></tr>
<tr class="separator:a7aedcc0352524f85eeabd6ea89b299c8 inherit pub_methods_class_thermodynamics"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af7ac1a0fb2cb4b91bdf95ac2bc6bb5aa inherit pub_methods_class_thermodynamics"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_thermodynamics.html#af7ac1a0fb2cb4b91bdf95ac2bc6bb5aa">ReloadDataTables</a> (const double new_temperature=-1.0)</td></tr>
<tr class="separator:af7ac1a0fb2cb4b91bdf95ac2bc6bb5aa inherit pub_methods_class_thermodynamics"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3b44af5d3530994d9cc06ad3b0c59152 inherit pub_methods_class_thermodynamics"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_thermodynamics.html#a3b44af5d3530994d9cc06ad3b0c59152">VerifyThermodynamic</a> ()</td></tr>
<tr class="memdesc:a3b44af5d3530994d9cc06ad3b0c59152 inherit pub_methods_class_thermodynamics"><td class="mdescLeft">&#160;</td><td class="mdescRight">Force the datatables to be read if they haven't already. Return true if the tables were already loaded or if the attempt to (re)open them succeded.  <a href="class_thermodynamics.html#a3b44af5d3530994d9cc06ad3b0c59152">More...</a><br /></td></tr>
<tr class="separator:a3b44af5d3530994d9cc06ad3b0c59152 inherit pub_methods_class_thermodynamics"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad523e6bf2d4c7531d13d65156bfb5b56 inherit pub_methods_class_thermodynamics"><td class="memItemLeft" align="right" valign="top">datatable *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_thermodynamics.html#ad523e6bf2d4c7531d13d65156bfb5b56">GetDatatable</a> ()</td></tr>
<tr class="separator:ad523e6bf2d4c7531d13d65156bfb5b56 inherit pub_methods_class_thermodynamics"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5d0a5c34a9accffd31325d71e73392e4 inherit pub_methods_class_thermodynamics"><td class="memItemLeft" align="right" valign="top">datatable *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_thermodynamics.html#a5d0a5c34a9accffd31325d71e73392e4">GetEnthalpyTable</a> (const char *alphabet=NULL)</td></tr>
<tr class="separator:a5d0a5c34a9accffd31325d71e73392e4 inherit pub_methods_class_thermodynamics"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9ea0b319beb59e9b6b3fa7b669140495 inherit pub_methods_class_thermodynamics"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_thermodynamics.html#a9ea0b319beb59e9b6b3fa7b669140495">ClearEnergies</a> ()</td></tr>
<tr class="memdesc:a9ea0b319beb59e9b6b3fa7b669140495 inherit pub_methods_class_thermodynamics"><td class="mdescLeft">&#160;</td><td class="mdescRight">Clear the currently loaded energy datatable and release its resources.  <a href="class_thermodynamics.html#a9ea0b319beb59e9b6b3fa7b669140495">More...</a><br /></td></tr>
<tr class="separator:a9ea0b319beb59e9b6b3fa7b669140495 inherit pub_methods_class_thermodynamics"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae093ea0bb4eb42f0cf29b79ebc6c5374 inherit pub_methods_class_thermodynamics"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_thermodynamics.html#ae093ea0bb4eb42f0cf29b79ebc6c5374">ClearEnthalpies</a> ()</td></tr>
<tr class="memdesc:ae093ea0bb4eb42f0cf29b79ebc6c5374 inherit pub_methods_class_thermodynamics"><td class="mdescLeft">&#160;</td><td class="mdescRight">Clear the currently loaded enthalpy datatable and release its resources.  <a href="class_thermodynamics.html#ae093ea0bb4eb42f0cf29b79ebc6c5374">More...</a><br /></td></tr>
<tr class="separator:ae093ea0bb4eb42f0cf29b79ebc6c5374 inherit pub_methods_class_thermodynamics"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a77601cfade7570057d46e5bd41a42cb3 inherit pub_methods_class_thermodynamics"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_thermodynamics.html#a77601cfade7570057d46e5bd41a42cb3">GetEnergyRead</a> () const</td></tr>
<tr class="memdesc:a77601cfade7570057d46e5bd41a42cb3 inherit pub_methods_class_thermodynamics"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return whether this instance of <a class="el" href="class_thermodynamics.html" title="Thermodynamics Class. ">Thermodynamics</a> has the paremters populated (either from disk or from another <a class="el" href="class_thermodynamics.html" title="Thermodynamics Class. ">Thermodynamics</a> class).  <a href="class_thermodynamics.html#a77601cfade7570057d46e5bd41a42cb3">More...</a><br /></td></tr>
<tr class="separator:a77601cfade7570057d46e5bd41a42cb3 inherit pub_methods_class_thermodynamics"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a24d39e4ec0a86aa68fa911147108827b inherit pub_methods_class_thermodynamics"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_thermodynamics.html#a24d39e4ec0a86aa68fa911147108827b">IsAlphabetRead</a> () const</td></tr>
<tr class="separator:a24d39e4ec0a86aa68fa911147108827b inherit pub_methods_class_thermodynamics"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adf2662bfd64db4a5b7c0cb2c2661440d inherit pub_methods_class_thermodynamics"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_thermodynamics.html#adf2662bfd64db4a5b7c0cb2c2661440d">~Thermodynamics</a> ()</td></tr>
<tr class="separator:adf2662bfd64db4a5b7c0cb2c2661440d inherit pub_methods_class_thermodynamics"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-methods"></a>
Static Public Member Functions</h2></td></tr>
<tr class="memitem:a939a33d1f65d21c257bbb2a0a2505953"><td class="memItemLeft" align="right" valign="top">static const char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953">GetErrorMessage</a> (const int error)</td></tr>
<tr class="memdesc:a939a33d1f65d21c257bbb2a0a2505953"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return error messages based on code from GetErrorCode and other error codes.  <a href="#a939a33d1f65d21c257bbb2a0a2505953">More...</a><br /></td></tr>
<tr class="separator:a939a33d1f65d21c257bbb2a0a2505953"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pro-methods"></a>
Protected Member Functions</h2></td></tr>
<tr class="memitem:a119769e10ee6b1846209a247643ae9f7"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a119769e10ee6b1846209a247643ae9f7">FileReader</a> (const char filename[], const <a class="el" href="_r_n_a_8h.html#a3883db120ec3ca0fe72f322a76dd2378">RNAInputType</a> fileType)</td></tr>
<tr class="separator:a119769e10ee6b1846209a247643ae9f7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4eb047c1e53a52c0ce515e0289e6e84f"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a4eb047c1e53a52c0ce515e0289e6e84f">init</a> (const char *sequenceOrFileName, const <a class="el" href="_r_n_a_8h.html#a3883db120ec3ca0fe72f322a76dd2378">RNAInputType</a> fileType, const bool allowUnknownBases=false, const bool <a class="el" href="class_thermodynamics.html#a018762c28a32e0a743b3a0ed5b808308">skipThermoTables</a>=false)</td></tr>
<tr class="separator:a4eb047c1e53a52c0ce515e0289e6e84f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="inherit_header pro_methods_class_thermodynamics"><td colspan="2" onclick="javascript:toggleInherit('pro_methods_class_thermodynamics')"><img src="closed.png" alt="-"/>&#160;Protected Member Functions inherited from <a class="el" href="class_thermodynamics.html">Thermodynamics</a></td></tr>
<tr class="memitem:af97dd7eac3218a8348c39366c725d190 inherit pro_methods_class_thermodynamics"><td class="memItemLeft" align="right" valign="top">virtual void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_thermodynamics.html#af97dd7eac3218a8348c39366c725d190">CopyThermo</a> (const <a class="el" href="class_thermodynamics.html">Thermodynamics</a> &amp;copy)</td></tr>
<tr class="memdesc:af97dd7eac3218a8348c39366c725d190 inherit pro_methods_class_thermodynamics"><td class="mdescLeft">&#160;</td><td class="mdescRight">Copy thermodynamic parameters from an instance of <a class="el" href="class_thermodynamics.html" title="Thermodynamics Class. ">Thermodynamics</a> class.  <a href="class_thermodynamics.html#af97dd7eac3218a8348c39366c725d190">More...</a><br /></td></tr>
<tr class="separator:af97dd7eac3218a8348c39366c725d190 inherit pro_methods_class_thermodynamics"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pro-attribs"></a>
Protected Attributes</h2></td></tr>
<tr class="memitem:acf869e72c097bae7824a72e8771e7a5a"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#acf869e72c097bae7824a72e8771e7a5a">ErrorCode</a></td></tr>
<tr class="separator:acf869e72c097bae7824a72e8771e7a5a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a16c4c8714246b97262acb50b4b2aad1d"><td class="memItemLeft" align="right" valign="top">ProgressHandler *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a16c4c8714246b97262acb50b4b2aad1d">progress</a></td></tr>
<tr class="separator:a16c4c8714246b97262acb50b4b2aad1d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a11556061f3292c11b40708b7cefb0b35"><td class="memItemLeft" align="right" valign="top">PFPRECISION *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a11556061f3292c11b40708b7cefb0b35">w5</a></td></tr>
<tr class="separator:a11556061f3292c11b40708b7cefb0b35"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abab25d8889f843589176b4ca2498fd20"><td class="memItemLeft" align="right" valign="top">PFPRECISION *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#abab25d8889f843589176b4ca2498fd20">w3</a></td></tr>
<tr class="separator:abab25d8889f843589176b4ca2498fd20"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a22519dbb4cb68fc353dfabc3a249d99c"><td class="memItemLeft" align="right" valign="top">PFPRECISION **&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a22519dbb4cb68fc353dfabc3a249d99c">wca</a></td></tr>
<tr class="separator:a22519dbb4cb68fc353dfabc3a249d99c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3ee830c6817a9a5da8216589b64a7b4d"><td class="memItemLeft" align="right" valign="top">pfdatatable *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a3ee830c6817a9a5da8216589b64a7b4d">pfdata</a></td></tr>
<tr class="separator:a3ee830c6817a9a5da8216589b64a7b4d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7c17cadb7eddf9094421a1bd0c7d22cb"><td class="memItemLeft" align="right" valign="top">DynProgArray&lt; PFPRECISION &gt; *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a7c17cadb7eddf9094421a1bd0c7d22cb">w</a></td></tr>
<tr class="separator:a7c17cadb7eddf9094421a1bd0c7d22cb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aec3ea67d1df58e7b3239f8bf03ed0c55"><td class="memItemLeft" align="right" valign="top">DynProgArray&lt; PFPRECISION &gt; *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#aec3ea67d1df58e7b3239f8bf03ed0c55">v</a></td></tr>
<tr class="separator:aec3ea67d1df58e7b3239f8bf03ed0c55"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7fac02aee6bfebae87d44941693ec66d"><td class="memItemLeft" align="right" valign="top">DynProgArray&lt; PFPRECISION &gt; *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a7fac02aee6bfebae87d44941693ec66d">wmb</a></td></tr>
<tr class="separator:a7fac02aee6bfebae87d44941693ec66d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa9df4fb4c33660eac913f67390c862e0"><td class="memItemLeft" align="right" valign="top">DynProgArray&lt; PFPRECISION &gt; *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#aa9df4fb4c33660eac913f67390c862e0">wl</a></td></tr>
<tr class="separator:aa9df4fb4c33660eac913f67390c862e0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2c62121fec224f1c2f6705534a34f91c"><td class="memItemLeft" align="right" valign="top">DynProgArray&lt; PFPRECISION &gt; *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a2c62121fec224f1c2f6705534a34f91c">wmbl</a></td></tr>
<tr class="separator:a2c62121fec224f1c2f6705534a34f91c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a208752cb6f6c11e0be87b8ab22038d93"><td class="memItemLeft" align="right" valign="top">DynProgArray&lt; PFPRECISION &gt; *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a208752cb6f6c11e0be87b8ab22038d93">wcoax</a></td></tr>
<tr class="separator:a208752cb6f6c11e0be87b8ab22038d93"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac72cbeed4c3618bf170c4afea4ce2d70"><td class="memItemLeft" align="right" valign="top">DynProgArray&lt; PFPRECISION &gt; *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ac72cbeed4c3618bf170c4afea4ce2d70">wlc</a></td></tr>
<tr class="separator:ac72cbeed4c3618bf170c4afea4ce2d70"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5701fc30b0ed83b60770f84f929e1909"><td class="memItemLeft" align="right" valign="top">PFPRECISION&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a5701fc30b0ed83b60770f84f929e1909">Q</a></td></tr>
<tr class="separator:a5701fc30b0ed83b60770f84f929e1909"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="inherit_header pro_attribs_class_thermodynamics"><td colspan="2" onclick="javascript:toggleInherit('pro_attribs_class_thermodynamics')"><img src="closed.png" alt="-"/>&#160;Protected Attributes inherited from <a class="el" href="class_thermodynamics.html">Thermodynamics</a></td></tr>
<tr class="memitem:ae2dd8a17877c6dff9e83174bc56b3c8a inherit pro_attribs_class_thermodynamics"><td class="memItemLeft" align="right" valign="top">datatable *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_thermodynamics.html#ae2dd8a17877c6dff9e83174bc56b3c8a">data</a></td></tr>
<tr class="separator:ae2dd8a17877c6dff9e83174bc56b3c8a inherit pro_attribs_class_thermodynamics"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a772a53fca8a3a103d9f8834980f2bc5f inherit pro_attribs_class_thermodynamics"><td class="memItemLeft" align="right" valign="top">datatable *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_thermodynamics.html#a772a53fca8a3a103d9f8834980f2bc5f">enthalpy</a></td></tr>
<tr class="separator:a772a53fca8a3a103d9f8834980f2bc5f inherit pro_attribs_class_thermodynamics"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab4f1b5c63568e04d7a45cccd67214b32 inherit pro_attribs_class_thermodynamics"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_thermodynamics.html#ab4f1b5c63568e04d7a45cccd67214b32">copied</a></td></tr>
<tr class="separator:ab4f1b5c63568e04d7a45cccd67214b32 inherit pro_attribs_class_thermodynamics"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae06b0fed0cc082d8d68238c5186319ec inherit pro_attribs_class_thermodynamics"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_thermodynamics.html#ae06b0fed0cc082d8d68238c5186319ec">nominal_temperature</a></td></tr>
<tr class="separator:ae06b0fed0cc082d8d68238c5186319ec inherit pro_attribs_class_thermodynamics"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab01edf6bb92a577eb8580940225be0d4 inherit pro_attribs_class_thermodynamics"><td class="memItemLeft" align="right" valign="top">string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_thermodynamics.html#ab01edf6bb92a577eb8580940225be0d4">nominal_alphabetName</a></td></tr>
<tr class="separator:ab01edf6bb92a577eb8580940225be0d4 inherit pro_attribs_class_thermodynamics"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a018762c28a32e0a743b3a0ed5b808308 inherit pro_attribs_class_thermodynamics"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_thermodynamics.html#a018762c28a32e0a743b3a0ed5b808308">skipThermoTables</a></td></tr>
<tr class="separator:a018762c28a32e0a743b3a0ed5b808308 inherit pro_attribs_class_thermodynamics"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pri-attribs"></a>
Private Attributes</h2></td></tr>
<tr class="memitem:a408ad69397beb1ffd831191114fd7630"><td class="memItemLeft" align="right" valign="top">structure *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a408ad69397beb1ffd831191114fd7630">ct</a></td></tr>
<tr class="separator:a408ad69397beb1ffd831191114fd7630"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a58544d6b4e93569ef51898208be06d37"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a58544d6b4e93569ef51898208be06d37">partitionfunctionallocated</a></td></tr>
<tr class="separator:a58544d6b4e93569ef51898208be06d37"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3c34e05e63dc1941784f1108b119019e"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a3c34e05e63dc1941784f1108b119019e">energyallocated</a></td></tr>
<tr class="separator:a3c34e05e63dc1941784f1108b119019e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0df9aab831d67cd88c2c0c95daef95bf"><td class="memItemLeft" align="right" valign="top">DynProgArray&lt; integersize &gt; *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a0df9aab831d67cd88c2c0c95daef95bf">ew2</a></td></tr>
<tr class="separator:a0df9aab831d67cd88c2c0c95daef95bf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8e958d96bdc5439caac1f1d7471ac0e9"><td class="memItemLeft" align="right" valign="top">DynProgArray&lt; integersize &gt; *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a8e958d96bdc5439caac1f1d7471ac0e9">ewmb2</a></td></tr>
<tr class="separator:a8e958d96bdc5439caac1f1d7471ac0e9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a43e0afefc5122174ea5f3065e58d0a29"><td class="memItemLeft" align="right" valign="top">integersize *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a43e0afefc5122174ea5f3065e58d0a29">ew5</a></td></tr>
<tr class="separator:a43e0afefc5122174ea5f3065e58d0a29"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a34ff98ed1d5627cfaed571447a4a71fb"><td class="memItemLeft" align="right" valign="top">integersize *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a34ff98ed1d5627cfaed571447a4a71fb">ew3</a></td></tr>
<tr class="separator:a34ff98ed1d5627cfaed571447a4a71fb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acf9f5a2f7a2f2142b0efe2224419f553"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#acf9f5a2f7a2f2142b0efe2224419f553">vmin</a></td></tr>
<tr class="separator:acf9f5a2f7a2f2142b0efe2224419f553"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ace814d16ca5fedb855c9aa0f6f2bc9b4"><td class="memItemLeft" align="right" valign="top">DynProgArray&lt; integersize &gt; *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ace814d16ca5fedb855c9aa0f6f2bc9b4">ev</a></td></tr>
<tr class="separator:ace814d16ca5fedb855c9aa0f6f2bc9b4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae265114cc811d410c0ea14a20a2fdb0b"><td class="memItemLeft" align="right" valign="top">DynProgArray&lt; integersize &gt; *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ae265114cc811d410c0ea14a20a2fdb0b">ew</a></td></tr>
<tr class="separator:ae265114cc811d410c0ea14a20a2fdb0b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a835a2d2ed7ac6e253920b916949bea46"><td class="memItemLeft" align="right" valign="top">DynProgArray&lt; integersize &gt; *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a835a2d2ed7ac6e253920b916949bea46">ewmb</a></td></tr>
<tr class="separator:a835a2d2ed7ac6e253920b916949bea46"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a96865031568d528c1e72ca7c77be6417"><td class="memItemLeft" align="right" valign="top">bool *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a96865031568d528c1e72ca7c77be6417">lfce</a></td></tr>
<tr class="separator:a96865031568d528c1e72ca7c77be6417"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad7a7258298a689bf42fc43a41a16effb"><td class="memItemLeft" align="right" valign="top">bool *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ad7a7258298a689bf42fc43a41a16effb">mod</a></td></tr>
<tr class="separator:ad7a7258298a689bf42fc43a41a16effb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4d886f6e92f47962bf2c43f4d24b52bd"><td class="memItemLeft" align="right" valign="top">forceclass *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a4d886f6e92f47962bf2c43f4d24b52bd">fce</a></td></tr>
<tr class="separator:a4d886f6e92f47962bf2c43f4d24b52bd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7dbc86bea8ba359871e6b4ad44ccc0a5"><td class="memItemLeft" align="right" valign="top">coordinates *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#a7dbc86bea8ba359871e6b4ad44ccc0a5">structurecoordinates</a></td></tr>
<tr class="separator:a7dbc86bea8ba359871e6b4ad44ccc0a5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae25e6cb1b35e47ab74890167fdc2cac4"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#ae25e6cb1b35e47ab74890167fdc2cac4">drawallocated</a></td></tr>
<tr class="separator:ae25e6cb1b35e47ab74890167fdc2cac4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adb078ada0b1e8fc227b448e62f5634d3"><td class="memItemLeft" align="right" valign="top">string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_r_n_a.html#adb078ada0b1e8fc227b448e62f5634d3">lastErrorDetails</a></td></tr>
<tr class="separator:adb078ada0b1e8fc227b448e62f5634d3"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="inherited"></a>
Additional Inherited Members</h2></td></tr>
<tr class="inherit_header pub_attribs_class_thermodynamics"><td colspan="2" onclick="javascript:toggleInherit('pub_attribs_class_thermodynamics')"><img src="closed.png" alt="-"/>&#160;Public Attributes inherited from <a class="el" href="class_thermodynamics.html">Thermodynamics</a></td></tr>
<tr class="memitem:ad088a23ca46d22bfd6e869f75e1ccac1 inherit pub_attribs_class_thermodynamics"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_thermodynamics.html#ad088a23ca46d22bfd6e869f75e1ccac1">isrna</a></td></tr>
<tr class="separator:ad088a23ca46d22bfd6e869f75e1ccac1 inherit pub_attribs_class_thermodynamics"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a id="a43195bcf410f9c53a0fb146fe490b6ea"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a43195bcf410f9c53a0fb146fe490b6ea">&#9670;&nbsp;</a></span>RNA() <span class="overload">[1/5]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">RNA::RNA </td>
          <td>(</td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>sequence</em>[], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const bool&#160;</td>
          <td class="paramname"><em>IsRNA</em> = <code>true</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Constructor - user provides a sequence as a c string. </p>
<p>Input sequence should contain A,C,G,T,U,a,c,g,t,u,x,X. The sequence is case-sensitive. Lower case letters are marked as unpaired. T=t=u=U. If IsRNA is true, the backbone is <a class="el" href="class_r_n_a.html">RNA</a>, so U is assumed. If IsRNA is false, the backbone is DNA, so T is assumed. x=X= nucleotide that neither stacks nor pairs. Unknown nucs result in an error. Note that sequences will subsequently be indexed starting at 1 (like a biologist), so that the 0th position in the sequence array will be nucleotide 1. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">sequence</td><td>is a NULL terminated c string. </td></tr>
    <tr><td class="paramname">IsRNA</td><td>is a bool that indicates whether this sequence is <a class="el" href="class_r_n_a.html">RNA</a> or DNA. true=<a class="el" href="class_r_n_a.html">RNA</a>. false=DNA. Default is true. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a92d94cbac25f6e96074c209ba9fa491b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a92d94cbac25f6e96074c209ba9fa491b">&#9670;&nbsp;</a></span>RNA() <span class="overload">[2/5]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">RNA::RNA </td>
          <td>(</td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>filepathOrSequence</em>[], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="_r_n_a_8h.html#a3883db120ec3ca0fe72f322a76dd2378">RNAInputType</a>&#160;</td>
          <td class="paramname"><em>fileType</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *const&#160;</td>
          <td class="paramname"><em>alphabetName</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const bool&#160;</td>
          <td class="paramname"><em>allowUnknownBases</em> = <code>false</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const bool&#160;</td>
          <td class="paramname"><em>skipThermoTables</em> = <code>false</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Constructor - user provides a filename for existing file as a c string. The existing file, specified by filename, can either be a ct file, a sequence, or an RNAstructure save file. This constructor generates internal error codes that can be accessed by <a class="el" href="class_r_n_a.html#a08a3be93861700dce517caf7ccf77358" title="Return an error code, where a return of zero is no error. ">GetErrorCode()</a> after the constructor is called. 0 = no error. The errorcode can be resolved to a c string using GetErrorMessage. Note that the contructor needs to be explicitly told, via IsRNA, what the backbone is because files do not store this information. Note also that save files explicitly store the thermodynamic parameters, therefore changing the backbone type as compared to the original calculation will not change structure predictions. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">filepathOrSequence</td><td>is null terminated c string that can be either a path to a CT, SEQ, FASTA, PFS, SAV, or DBN file or a full <a class="el" href="class_r_n_a.html">RNA</a> sequence (whitepsace allowed) </td></tr>
    <tr><td class="paramname">fileType</td><td>is an RNAInputType (e.g FILE_CT, FILE_SEQ etc) that indicates what type of file filepathOrSequence refers to. If filepathOrSequence is an in-place sequence, fileType should be SEQUENCE_STRING. </td></tr>
    <tr><td class="paramname">alphabetName</td><td>is a c-string that specifies the alphabet name (e.g. "rna", "dna", etc) </td></tr>
    <tr><td class="paramname">allowUnknownBases</td><td>boolean that determines the handling of unknown bases (i.e. those not defined in the alphabet) If allowUnknownBases is true, they will be interpreted as bases that neither pair nor stack. Otherwise they will result in an error. </td></tr>
    <tr><td class="paramname">skipThermoTables</td><td>If true, only the alphabet specification will be loaded, not the full thermodynamic energy tables. The default is false &ndash; i.e. load all tables. This should only be set to true in programs that never make use of energies &ndash; i.e. ct2dot etc. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a9e572b36b07e0b2fee3165930a430472"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9e572b36b07e0b2fee3165930a430472">&#9670;&nbsp;</a></span>RNA() <span class="overload">[3/5]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">RNA::RNA </td>
          <td>(</td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>filepathOrSequence</em>[], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="_r_n_a_8h.html#a3883db120ec3ca0fe72f322a76dd2378">RNAInputType</a>&#160;</td>
          <td class="paramname"><em>fileType</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="class_thermodynamics.html">Thermodynamics</a> *&#160;</td>
          <td class="paramname"><em>copyThermo</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Constructor - user provides a filename for existing file as a c string. Construct an <a class="el" href="class_r_n_a.html">RNA</a> from either a sequence or file. Copy the <a class="el" href="class_thermodynamics.html" title="Thermodynamics Class. ">Thermodynamics</a> info and datatable from another <a class="el" href="class_thermodynamics.html" title="Thermodynamics Class. ">Thermodynamics</a> instance. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">filepathOrSequence</td><td>is null terminated c string that can be either a path to a CT, SEQ, FASTA, PFS, SAV, or DBN file or a full <a class="el" href="class_r_n_a.html">RNA</a> sequence (whitepsace allowed) </td></tr>
    <tr><td class="paramname">fileType</td><td>is an RNAInputType (e.g FILE_CT, FILE_SEQ etc) that indicates what type of file filepathOrSequence refers to. If filepathOrSequence is an in-place sequence, fileType should be SEQUENCE_STRING. </td></tr>
    <tr><td class="paramname">copyThermo</td><td>A pointer to an existing <a class="el" href="class_thermodynamics.html" title="Thermodynamics Class. ">Thermodynamics</a> from which to copy the datatables. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a800274d25e5fe015b5672c6e7ed8214e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a800274d25e5fe015b5672c6e7ed8214e">&#9670;&nbsp;</a></span>RNA() <span class="overload">[4/5]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">RNA::RNA </td>
          <td>(</td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>filename</em>[], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="_r_n_a_8h.html#a3883db120ec3ca0fe72f322a76dd2378">RNAInputType</a>&#160;</td>
          <td class="paramname"><em>fileType</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const bool&#160;</td>
          <td class="paramname"><em>IsRNA</em> = <code>true</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Constructor - user provides a filename for existing file as a c string. </p>
<p>The existing file, specified by filename, can either be a ct file, a sequence, or an RNAstructure save file. Therefore, the user provides a flag for the file: 1 =&gt; .ct file, 2 =&gt; .seq file, 3 =&gt; partition function save (.pfs) file, 4 =&gt; folding save file (.sav), 5 =&gt; DotBracket (DBN) file This constructor generates internal error codes that can be accessed by <a class="el" href="class_r_n_a.html#a08a3be93861700dce517caf7ccf77358" title="Return an error code, where a return of zero is no error. ">GetErrorCode()</a> after the constructor is called. 0 = no error. The errorcode can be resolved to a c string using GetErrorMessage. Note that the contructor needs to be explicitly told, via IsRNA, what the backbone is because files do not store this information. Note also that save files explicitly store the thermodynamic parameters, therefore changing the backbone type as compaared to the original calculation will not change structure predictions. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">filename</td><td>is null terminated c string. </td></tr>
    <tr><td class="paramname">fileType</td><td>is an integer that indicates the file type. </td></tr>
    <tr><td class="paramname">IsRNA</td><td>is a bool that indicates whether this sequence is <a class="el" href="class_r_n_a.html">RNA</a> or DNA. true=<a class="el" href="class_r_n_a.html">RNA</a>. false=DNA. Default is true. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="aa1ab0c1a0065091f8fdf41c2cdfb43ea"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa1ab0c1a0065091f8fdf41c2cdfb43ea">&#9670;&nbsp;</a></span>RNA() <span class="overload">[5/5]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">RNA::RNA </td>
          <td>(</td>
          <td class="paramtype">const bool&#160;</td>
          <td class="paramname"><em>IsRNA</em> = <code>true</code></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Default Constructor - user provides nothing. This basic constructor is provided for bimolecular folding and should not generally need to be accessed by end users of the <a class="el" href="class_r_n_a.html">RNA</a> class. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">IsRNA</td><td>is a bool that indicates whether this sequence is <a class="el" href="class_r_n_a.html">RNA</a> or DNA. true=<a class="el" href="class_r_n_a.html">RNA</a>. false=DNA. Default is true. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="adf930d0e1dd46d4d7dedf006228e2bbf"></a>
<h2 class="memtitle"><span class="permalink"><a href="#adf930d0e1dd46d4d7dedf006228e2bbf">&#9670;&nbsp;</a></span>~RNA()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">RNA::~RNA </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Destructor. </p>
<p>The destructor automatically cleans up all allocated memory for predicted or specified structures. </p>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="a1eebbd7916abda62860c611e04bd977d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1eebbd7916abda62860c611e04bd977d">&#9670;&nbsp;</a></span>AddComment()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::AddComment </td>
          <td>(</td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>comment</em>[], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>structurenumber</em> = <code>1</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Add a comment associated with a structure. </p>
<p>This comment will appear in a written .ct file. The comment is appended to any existing comments, like titles read from .seq files. This function is especially useful if the constructor is used in which a character array is provided with the sequence. In that case, there is no sequence title read. The function returns 0 in the case of no errors, or 3 if the structurenumber is invalid. An error message can be retrieved using <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> called with the errorcode. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">comment</td><td>is a character array that contains a null terminated c-string with the comment to be registered. </td></tr>
    <tr><td class="paramname">structurenumber</td><td>is an integer that specifies to which structure the comment should be added. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that contains an error code, where 0 is no error and non-zero is an error. </dd></dl>

</div>
</div>
<a id="ae052a7062e9eb11c74027eba2f41d5d7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae052a7062e9eb11c74027eba2f41d5d7">&#9670;&nbsp;</a></span>BreakPseudoknot()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::BreakPseudoknot </td>
          <td>(</td>
          <td class="paramtype">const bool&#160;</td>
          <td class="paramname"><em>minimum_energy</em> = <code>true</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>structurenumber</em> = <code>0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const bool&#160;</td>
          <td class="paramname"><em>useFastMethod</em> = <code>true</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Break any pseudoknots that might be in a structure. </p>
<p>This function uses the method of Smit et al. to break pseudoknots by running a dynamic programming algorithm that cannot predict pseudoknots while only allowing the pairs that already exist in the structure. When minimum_energy = true (the default), this function predicts the lowest free energy structure that has no pseudoknots. Note that when minimum_energy = true, this function might additionally break pairs that are not pseudoknotted if the pairs increase the folding free energy change or are forbidden (as in an isolated pair). Also note that when minum_energy=true, this function uses the GenerateAllSuboptimalStructures methodology behind the scenes, so large internal loops in the input would lead to a loss of pairs. When minumum_energy is set to false, this function maximizes the number of base pairs in the pseudoknot free structure. Return 0 if no error and non-zero errors can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">minimum_energy</td><td>is a bool thgat indicates where the structure should be minimum in free energy (true) or maximize pairs (false). </td></tr>
    <tr><td class="paramname">structurenumber</td><td>is an int that indicates a specific structure for which to break pseudoknots (indexed from 1). The default value, 0, indicates that all structures should have pseudoknots broken. </td></tr>
    <tr><td class="paramname">useFastMethod</td><td>If true (default) a fast method that is energy and sequence agnostic will be used to find pseudoknots. This method should be identical in results to MEAFill, but much faster. However, if minimum_energy is also true, AllTrace will be used instead of the fast method (i.e. this parameter will be ignored). </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>an int that provides an error code. 0 = no error. </dd></dl>

</div>
</div>
<a id="a815bd1662622d7588a80107777e713be"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a815bd1662622d7588a80107777e713be">&#9670;&nbsp;</a></span>CalculateFreeEnergy()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double RNA::CalculateFreeEnergy </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>structurenumber</em> = <code>1</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const bool&#160;</td>
          <td class="paramname"><em>UseSimpleMBLoopRules</em> = <code>false</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Return the predicted Gibb's free energy change for structure # structurenumber, defaulted to 1. </p>
<p>Free energies are in kcal/mol. The first time this is called, if no other free energy calculation has been performed and the folding temperature has not been specifed, thermodynamic parameter files (.dat) files will be read from disk. The parameter files should be located in the directory specified by environment variable $DATAPATH, or the pwd. In case of error, the function returns a free energy change of zero. Note!: That a free energy change of zero is also a valid folding free energy change. Errors will also generate an internal error code, accessible with <a class="el" href="class_r_n_a.html#a08a3be93861700dce517caf7ccf77358" title="Return an error code, where a return of zero is no error. ">GetErrorCode()</a>. <a class="el" href="class_r_n_a.html#a08a3be93861700dce517caf7ccf77358" title="Return an error code, where a return of zero is no error. ">GetErrorCode()</a> will return 0 when there is no error and other codes can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">structurenumber</td><td>is an integer that refers to the index of the structure for which to calculate the folding free energy change. This defaults to 1. </td></tr>
    <tr><td class="paramname">UseSimpleMBLoopRules</td><td>is a bool that indicates what energy rules to use. The default, false, uses the complete nearest neighbor model for multibranch loops. When true is passed, the energy model is instead a simplified model that is the one used by the dynamic programming algorithms. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A double which is the folding free energy change in kcal/mol. </dd></dl>

</div>
</div>
<a id="abee55f46e52b836f32c9a572af8e2d7c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abee55f46e52b836f32c9a572af8e2d7c">&#9670;&nbsp;</a></span>ContainsPseudoknot()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool RNA::ContainsPseudoknot </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>structurenumber</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Report if there are any pseudoknots in a structure. </p>
<p>This method checks for any "crossing pairs," i.e. i-j and i'-j' s.t. i &lt; i' &lt; j &lt; jp. If there is at least one crossing pair set, then there is a pseudoknot and the function returns true. This function generates internal error codes that can be accessed by <a class="el" href="class_r_n_a.html#a08a3be93861700dce517caf7ccf77358" title="Return an error code, where a return of zero is no error. ">GetErrorCode()</a> after the constructor is called: 0 = no error, nonzero = error. The errorcode can be resolved to a c string using GetErrorMessage. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">structurenumber</td><td>is an int that indicates the structure number to check. Note that indexing of structures starts with structure #1. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A bool that indicates whether there is a pseudoknot. </dd></dl>

</div>
</div>
<a id="a155d2c6ae7d094206ec0699228f67c46"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a155d2c6ae7d094206ec0699228f67c46">&#9670;&nbsp;</a></span>CopyThermo()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void RNA::CopyThermo </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="class_thermodynamics.html">Thermodynamics</a> &amp;&#160;</td>
          <td class="paramname"><em>copy</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Copy thermodynamic parameters from an instance of an <a class="el" href="class_r_n_a.html">RNA</a> class. </p>
<p>This is generally not needed because functions automatically populate the parameters from disk. It is helpful, however, in constructors of derived classes. Normally the Thermodyanamic(Thermodynamic* copy) constructor should be used instead. Reason: Since the extended alphabet, the datatable is loaded in the <a class="el" href="class_r_n_a.html">RNA</a> class constructor, so copying the datatable must also happen in the constructor to avoid reading a (possibly different) datatable first. </p>

</div>
</div>
<a id="ae464f8408d0db524408ab27cd6ed5f6e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae464f8408d0db524408ab27cd6ed5f6e">&#9670;&nbsp;</a></span>DetermineDrawingCoordinates()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::DetermineDrawingCoordinates </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>height</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>width</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>structurenumber</em> = <code>1</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Determine the coordinates for drawing a secondary structure. </p>
<p>This function determines drawing coordinates for all nucleotides in structure number structurenumber. User must specify the height and width of a character (use the largest of nucleotides). The coordinates are in an abstract palette; the user must determine the minimum and maximum coordinate in both the x and y direction. The actual coordinates are fetched using <a class="el" href="class_r_n_a.html#a97c3e9a51d32a73fadd53f467c8f3ae0" title="Get the X coordinate for nucleotide i for drawing a structure. ">GetNucleotideXCoordinate(int i)</a> and <a class="el" href="class_r_n_a.html#acb2539f11b34a8e34668697ebf2890fb" title="Get the Y coordinate for nucleotide i for drawing a structure. ">GetNucleotideYCoordinate(int i)</a>. This function returns are error code, where 0=no error and other messages can be resolved to a c string using GetErrorMessage. The structure to be drawn must be free of pseudoknots. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">height</td><td>is an integer that refers to the height of a nucleotide. </td></tr>
    <tr><td class="paramname">width</td><td>is an integer that refers to the width of a nucleotide, where the largest nucleotide should be provided or a non-proportional font should be used. </td></tr>
    <tr><td class="paramname">structurenumber</td><td>is an integer that refers to the structure to be drawn. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An int that provides an error code, 0 = no error and other errors can be resolved to a c string using GetErrorMessage. </dd></dl>

</div>
</div>
<a id="afdca3d7732fbffe48dfc7d52f193798b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#afdca3d7732fbffe48dfc7d52f193798b">&#9670;&nbsp;</a></span>EnsureStructureCapcacity()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void RNA::EnsureStructureCapcacity </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>minimumStructures</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Ensure that at a minumum number of structures have been created. </p>

</div>
</div>
<a id="aa61adf54c3d44cc6e29f6ae523342491"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa61adf54c3d44cc6e29f6ae523342491">&#9670;&nbsp;</a></span>ExteriorLoopCorrection()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double RNA::ExteriorLoopCorrection </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>structurenumber</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const bool&#160;</td>
          <td class="paramname"><em>UseSimpleMBLoopRules</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>min_index</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>max_index</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a119769e10ee6b1846209a247643ae9f7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a119769e10ee6b1846209a247643ae9f7">&#9670;&nbsp;</a></span>FileReader()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::FileReader </td>
          <td>(</td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>filename</em>[], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="_r_n_a_8h.html#a3883db120ec3ca0fe72f322a76dd2378">RNAInputType</a>&#160;</td>
          <td class="paramname"><em>fileType</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="aaa9563532ecfeee0de773396b8fc9e17"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aaa9563532ecfeee0de773396b8fc9e17">&#9670;&nbsp;</a></span>FoldSingleStrand()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::FoldSingleStrand </td>
          <td>(</td>
          <td class="paramtype">const float&#160;</td>
          <td class="paramname"><em>percent</em> = <code>20</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>maximumstructures</em> = <code>20</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>window</em> = <code>5</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>savefile</em>[] = <code>&quot;&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>maxinternalloopsize</em> = <code>30</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>mfeonly</em> = <code>false</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>simple_iloops</em> = <code>true</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>disablecoax</em> = <code>false</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>allow_isolated</em> = <code>false</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Predict the lowest free energy secondary structure and generate suboptimal structures using a heuristic. </p>
<p>This function predicts the lowest free energy structure and suboptimal structures. If the temperature has not been specified using SetTemperature and no free energies have been calculated, the thermodynamic parameters have not been read and therefore they will be read by this function call. The parameter files should be located in the directory specified by the environment variable $DATAPATH of the pwd. In case of error, the function returns a non-zero that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">percent</td><td>is the maximum % difference in free energy in suboptimal structures from the lowest free energy structure. The default is 20. </td></tr>
    <tr><td class="paramname">maximumstructures</td><td>is the maximum number of suboptimal structures to generate. The default is 20. </td></tr>
    <tr><td class="paramname">window</td><td>is a parameter that specifies how different the suboptimal structures should be from each other (0=no restriction and larger integers require structures to be more different). The defaults is 5, but this should be customized based on sequence length. </td></tr>
    <tr><td class="paramname">savefile</td><td>is c string containing a file path and name for a savefile (.sav)that can be used to generate energy dot plots and to refold the secondary structure using different suboptimal structure parameters. The default is "", which results in no save file written. </td></tr>
    <tr><td class="paramname">maxinternalloopsize</td><td>is the maximum number of unpaired nucleotides in bulge and internal loops. This is used to accelerate the prediction speed. The default is 30. </td></tr>
    <tr><td class="paramname">mfeonly</td><td>is a bool that indicates whether only the minimum free energy structure will be generated. This saves half the calculation time, but no save file can be generated. Default is false. </td></tr>
    <tr><td class="paramname">allow_isolated</td><td>is a bool that indicates whether isolated base pairs are allowed. The default, false, is to use a heuristic to prevent isolated base pairs. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An int that indicates an error code (0 = no error, 5 = error reading thermodynamic parameter files, 14 = traceback error). </dd></dl>

</div>
</div>
<a id="a8ba352350bcda5514d772db7e763229e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8ba352350bcda5514d772db7e763229e">&#9670;&nbsp;</a></span>ForceDoubleStranded()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::ForceDoubleStranded </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>i</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Force a nucleotide to be double stranded (base paired). </p>
<p>This function indicates a nucleotide that is double stranded (paired). In subsequent structure prediction, this nucleotide will be double stranded. The function returns 0 with no error and a non-zero otherwise that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i</td><td>is the index of the paired nucleotide. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates an error code (0 = no error, 4 = nucleotide out of range, 8 = too many restraints specified, 9 = same nucleotide in conflicting restraint). </dd></dl>

</div>
</div>
<a id="ae2ecc054cf728366d884ecf8f066ee20"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae2ecc054cf728366d884ecf8f066ee20">&#9670;&nbsp;</a></span>ForceFMNCleavage()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::ForceFMNCleavage </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>i</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Indicate a nucleotide that is accessible to FMN cleavage (a U in GU pair). </p>
<p>In subsequent structure prediction, this nucleotide will be in a GU pair. The function returns 0 with no error and a non-zero otherwise that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i</td><td>is the index of the FMN-cleaved nucleotide. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates an error code (0 = no error, 4 = nucleotide out of range, 8 = too many restraints specified, 9 = same nucleotide in conflicting restraint, 11 = nucleotide not U). </dd></dl>

</div>
</div>
<a id="a99f37cbc69b7ee0e99116c6fe074e7a7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a99f37cbc69b7ee0e99116c6fe074e7a7">&#9670;&nbsp;</a></span>ForceMaximumPairingDistance()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::ForceMaximumPairingDistance </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>distance</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Force a maximum distance between apired nucleotides. </p>
<p>In a subsequent structure prediction, there will be no pairs allowed between nucleotides more distant than distance, i.e. |j-i| &lt; distance for i to pair to j. The function returns and error code; 0==no error, 12== too long or too short distance. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">distance</td><td>is the maximum pairing distance. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates an error code (0 = no error, 12 = too short). </dd></dl>

</div>
</div>
<a id="ac6f531c36afbcbb44dfad283a31c5879"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac6f531c36afbcbb44dfad283a31c5879">&#9670;&nbsp;</a></span>ForceModification()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::ForceModification </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>i</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Force modification for a nucleotide. </p>
<p>This function indicates a nucleotide that is accessible to chemical modification. In subsequent structure prediction, this nucleotide will be single stranded, at the end of a helix, or in or adjacent to a GU pair. The function returns 0 with no error and a non-zero otherwise that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i</td><td>is the index of the nucleotide accessible to chemical modification. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates an error code (0 = no error, 4 = nucleotide out of range, 8 = too many restraints specified). </dd></dl>

</div>
</div>
<a id="a6d33a787f66bbef7e78cf64624c0e12e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6d33a787f66bbef7e78cf64624c0e12e">&#9670;&nbsp;</a></span>ForcePair()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::ForcePair </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>i</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>j</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Force a pair between two nucleotides. </p>
<p>This function forces a pair between two nucleotides in subsequent structure predictions. When multiple pairs are specified, the pairs must not force a pseudoknot. The function returns 0 with no error and a non-zero otherwise that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i</td><td>is the index of one nucleotide in the pair. </td></tr>
    <tr><td class="paramname">j</td><td>is the index of the second nucleotide in the pair. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates an error code (0 = no error, 4 = nucleotide out of range, 6 = pseudoknot formation, 7 = non-canonical pair, 8 = too many restraints specified, 9 = same nucleotide in conflicting restraint). </dd></dl>

</div>
</div>
<a id="a4fa207c587f788981d860f70616055e4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4fa207c587f788981d860f70616055e4">&#9670;&nbsp;</a></span>ForceProhibitPair()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::ForceProhibitPair </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>i</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>j</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prohibit a pair between two nucleotides. </p>
<p>This function prevents a pair between two nucleotides in subsequent structure predictions. The function returns 0 with no error and a non-zero otherwise that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i</td><td>is the index of one nucleotide in the pair. </td></tr>
    <tr><td class="paramname">j</td><td>is the index of the second nucleotide in the pair. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates an error code (0 = no error, 4 = nucleotide out of range, 8 = too many restraints specified, 9 = nucleotide in conflicting restraint). </dd></dl>

</div>
</div>
<a id="aef005a05b220b79044a620d202aa8d1b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aef005a05b220b79044a620d202aa8d1b">&#9670;&nbsp;</a></span>ForceSingleStranded()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::ForceSingleStranded </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>i</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Force a nucleotide to be single stranded. </p>
<p>This function indicates a nucleotide that is single stranded. In subsequent structure prediction, this nucleotide will be single stranded. The function returns 0 with no error and a non-zero otherwise that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i</td><td>is the index of the nucleotide that is single stranded. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates an error code (0 = no error, 4 = nucleotide out of range, 8 = too many restraints specified, 9 = same nucleotide in conflicting restraint). </dd></dl>

</div>
</div>
<a id="ae9c16cf6e875e5152eca3f4b94721fe3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae9c16cf6e875e5152eca3f4b94721fe3">&#9670;&nbsp;</a></span>GenerateAllSuboptimalStructures()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::GenerateAllSuboptimalStructures </td>
          <td>(</td>
          <td class="paramtype">const float&#160;</td>
          <td class="paramname"><em>percent</em> = <code>5</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const double&#160;</td>
          <td class="paramname"><em>deltaG</em> = <code>0.6</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Predict the lowest free energy secondary structure and generate all suboptimal structures. </p>
<p>This function predicts the lowest free energy structure and suboptimal structures. If the temperature has not been specified using SetTemperature and no free energies have been calculated, the thermodynamic parameters have not been read and therefore they will be read by this function call. The parameter files should be located in the directory specified by the environment variable $DATAPATH of the pwd. In case of error, the function returns a non-zero that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. Two controls are available for limiting the number of structures, the maximum % difference in energy (percent) and the maximum absolute change in energy (deltaG). The smaller of the two will be used as the limit. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">percent</td><td>is the maximum % difference in free energy in suboptimal structures from the lowest free energy structure. The default is 5. </td></tr>
    <tr><td class="paramname">deltaG</td><td>is the maximum difference in free energy change above the lowest free energy structure (in kcal/mol). The defaults is 0.6 kcal/mol. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An int that indicates an error code (0 = no error, non-zero = error). </dd></dl>

</div>
</div>
<a id="a571a75afbe286a928b1910b19603273d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a571a75afbe286a928b1910b19603273d">&#9670;&nbsp;</a></span>GetBackboneType()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool RNA::GetBackboneType </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the backbone type. </p>
<p>This function returns whether the backbone is <a class="el" href="class_r_n_a.html">RNA</a> or DNA. Note that backbone type is set when calling the constructor. </p><dl class="section return"><dt>Returns</dt><dd>A bool that indicates the backbone (true = <a class="el" href="class_r_n_a.html">RNA</a>, false = DNA). </dd></dl>

</div>
</div>
<a id="ab329d510a2ce6765348142c8a6f7dcaa"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab329d510a2ce6765348142c8a6f7dcaa">&#9670;&nbsp;</a></span>GetCommentString()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::string RNA::GetCommentString </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>structurenumber</em> = <code>1</code></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Provide the comment from the ct file as a string. </p>
<p>This function provides the comment from the CT file for a structure as a string. This function generates internal error codes that can be accessed by <a class="el" href="class_r_n_a.html#a08a3be93861700dce517caf7ccf77358" title="Return an error code, where a return of zero is no error. ">GetErrorCode()</a> after the function is called: 0 = no error, nonzero = error. The errorcode can be resolved to a c string using GetErrorMessage. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">structurenumber</td><td>is the structure for which the comment is to be provided. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A string that provides the comment. </dd></dl>

</div>
</div>
<a id="a08789d87f4a205b16f643696206c5eb4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a08789d87f4a205b16f643696206c5eb4">&#9670;&nbsp;</a></span>GetEnsembleDefect()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double RNA::GetEnsembleDefect </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>structurenumber</em> = <code>1</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>start</em> = <code>0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>end</em> = <code>0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>filename</em>[] = <code>&quot;&quot;</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the ensemble defect of a secondary structure. </p>
<p>Returns the ensemble defect as determined by the partition function. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">structurenumber</td><td>is an integer indicating the predicted structure number. </td></tr>
    <tr><td class="paramname">start</td><td>is the start nucleotide (1-indexed) for a local calculation. </td></tr>
    <tr><td class="paramname">end</td><td>is the end nucleotide for a local calculation. </td></tr>
    <tr><td class="paramname">filename</td><td>is the optional name for a plain text file that will show the per nucleotide defect. NOTE that this file is appended to. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A double that is the ensemble defect. To get the Normalized Ensemble Defect, this number must be divided by the Sequence Length. </dd></dl>

</div>
</div>
<a id="a88e712e84b77792e813d22b7a51779f1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a88e712e84b77792e813d22b7a51779f1">&#9670;&nbsp;</a></span>GetEnsembleEnergy()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double RNA::GetEnsembleEnergy </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the ensemble folding free energy change. </p>
<p>Returns the ensemble folding free energy change as determined by the partition function. This is a handy way of getting the size of the partition function Q, which itself is too large to fit in a double for all but the shortest sequences. The ensemble folding free energy change = -RT ln (Q). Function requires that the partition function data be present either because <a class="el" href="class_r_n_a.html#a654db191fe8a0a76350dfc57b2c5740c" title="Predict the partition function for a sequence. ">PartitionFunction()</a> has been called or the constructor that reads a partition function save was used. This function generates internal error codes that can be accessed by <a class="el" href="class_r_n_a.html#a08a3be93861700dce517caf7ccf77358" title="Return an error code, where a return of zero is no error. ">GetErrorCode()</a> after the constructor is called: 0 = no error, nonzero = error. The errorcode can be resolved to a c string using GetErrorMessage. </p><dl class="section return"><dt>Returns</dt><dd>A double that is the ensemble folding free energy change in kcal/mol. </dd></dl>

</div>
</div>
<a id="a08a3be93861700dce517caf7ccf77358"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a08a3be93861700dce517caf7ccf77358">&#9670;&nbsp;</a></span>GetErrorCode()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::GetErrorCode </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Return an error code, where a return of zero is no error. </p>
<p>This function returns and error flag that is generated during construction by <a class="el" href="class_r_n_a.html">RNA</a>(const char &amp;filename, const int type, const bool IsRNA=true) or from <a class="el" href="class_r_n_a.html#a815bd1662622d7588a80107777e713be" title="Return the predicted Gibb&#39;s free energy change for structure # structurenumber, defaulted to 1...">CalculateFreeEnergy()</a>. An error of zero is always no error. Other codes are errors and a c-string can be fetched for the error with <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a>. </p><dl class="section return"><dt>Returns</dt><dd>An integer that provides the error code. </dd></dl>

</div>
</div>
<a id="a29ba3acd2c1e8ed61e8e0c367cc6051c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a29ba3acd2c1e8ed61e8e0c367cc6051c">&#9670;&nbsp;</a></span>GetErrorDetails()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">const string RNA::GetErrorDetails </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns extended details about the last error. (e.g. error messages produced during file read operations that are otherwise lost.) </p>

</div>
</div>
<a id="a939a33d1f65d21c257bbb2a0a2505953"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a939a33d1f65d21c257bbb2a0a2505953">&#9670;&nbsp;</a></span>GetErrorMessage()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">const char * RNA::GetErrorMessage </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>error</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Return error messages based on code from GetErrorCode and other error codes. </p>
<p>0 = no error 1 = input file not found 2 = error opening file 3 = structure number out of range 4 = nucleotide number out of range 5 = error reading thermodynamic parameters 6 = pseudoknot formation 7 = non-canonical pair 8 = too many restraints specified 9 = same nucleotide in conflicting restraint 10 = no structures to write 11 = nucleotide not a U (caused by <a class="el" href="class_r_n_a.html#ae2ecc054cf728366d884ecf8f066ee20" title="Indicate a nucleotide that is accessible to FMN cleavage (a U in GU pair). ">ForceFMNCleavage()</a> 12 = distance too short 13 = error reading constraint file 14 = traceback error 15 = no partition function data present 16 = incorrect save file version used 17 = cannot be performed without having read a save file (.sav) 18 = threshold is too low to be valid 19 = drawing coordinates have not been determined 20 = no sequence has been read 21 = over 1 probability error on stochastic traceback 22 = programming error, unrecognized input to constructor 23 = no structures present 24 = too few iterations 25 = index (for drawing) is not a multiple of 10 </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">error</td><td>is the integer error code provided by <a class="el" href="class_r_n_a.html#a08a3be93861700dce517caf7ccf77358" title="Return an error code, where a return of zero is no error. ">GetErrorCode()</a> or from other functions that return integer error codes. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A pointer to a c string that provides an error message. </dd></dl>

</div>
</div>
<a id="a3e1f322c241c729f36b56cb0e452c0c4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3e1f322c241c729f36b56cb0e452c0c4">&#9670;&nbsp;</a></span>GetErrorMessageString()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">string RNA::GetErrorMessageString </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>error</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Return error messages based on code from GetErrorCode and other error codes. </p>
<p>Although <a class="el" href="class_r_n_a.html">RNA</a> generally uses c strings, this function returns a string that is suitable for interfacing with JAVA, etc. See the error list in the <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> entry. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">error</td><td>is the integer error code provided by <a class="el" href="class_r_n_a.html#a08a3be93861700dce517caf7ccf77358" title="Return an error code, where a return of zero is no error. ">GetErrorCode()</a> or from other functions that return integer error codes. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A string that provides an error message. </dd></dl>

</div>
</div>
<a id="a7e1d3fc6f19b4f5c8fe5373763996029"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7e1d3fc6f19b4f5c8fe5373763996029">&#9670;&nbsp;</a></span>GetForcedDoubleStranded()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::GetForcedDoubleStranded </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>constraintnumber</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Return a nucleotide that is forced double stranded. </p>
<p>This function returns a nucleotide that is constrainted to be paired. Constraints are numbered from zero to GetNumberofForcedDoubleStranded()-1. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">constraintnumber</td><td>is the index to the constraint number. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that is the nucleotide index. If the constraintnumber is for a constraint that does not exist, zero is returned. </dd></dl>

</div>
</div>
<a id="a85711931f3d25f6c2f209ec64beba01c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a85711931f3d25f6c2f209ec64beba01c">&#9670;&nbsp;</a></span>GetForcedFMNCleavage()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::GetForcedFMNCleavage </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>constraintnumber</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Return a nucleotide that is accessible to FMN cleavage. </p>
<p>This function returns a nucleotide that is constrainted to be accessible to FMN cleavage (a U in a GU pair). Constraints are numbered from zero to GetNumberofForcedFMNCleavages()-1. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">constraintnumber</td><td>is the index to the constraint number. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that is the nucleotide index. If the constraintnumber is for a constraint that does not exist, zero is returned. </dd></dl>

</div>
</div>
<a id="a9c1eb7ad1992c12d69979418a14ba883"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9c1eb7ad1992c12d69979418a14ba883">&#9670;&nbsp;</a></span>GetForcedModification()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::GetForcedModification </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>constraintnumber</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Return a nucleotide that is accessible to modification. </p>
<p>This function returns a nucleotide that is constrainted to be accessible to chemical modification. Constraints are numbered from zero to GetNumberofModifications()-1. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">constraintnumber</td><td>is the index to the constraint number. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that is the nucleotide index. If the constraintnumber is for a constraint that does not exist, zero is returned. </dd></dl>

</div>
</div>
<a id="a1d459e60cbbd1d9648a3b61189c76c8d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1d459e60cbbd1d9648a3b61189c76c8d">&#9670;&nbsp;</a></span>GetForcedPair()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::GetForcedPair </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>constraintnumber</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const bool&#160;</td>
          <td class="paramname"><em>fiveprime</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Return a nucleotide in a forced pair. </p>
<p>This function returns either the five prime or three prime nucleotide in a forced pair constraint, depending on the value of fiveprime. Constraints are numbered from zero to GetNumberofForcedPairs()-1. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">constraintnumber</td><td>is the index to the constraint number. </td></tr>
    <tr><td class="paramname">fiveprime</td><td>determines if the nucleotide is the five prime or the three prime nucleotide in the constraint. true = five prime nucleotide. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that is the nucleotide index. If the constraintnumber is for a constraint that does not exist, zero is returned. </dd></dl>

</div>
</div>
<a id="ad993910aa7abba8e80cf3cbb590bddab"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad993910aa7abba8e80cf3cbb590bddab">&#9670;&nbsp;</a></span>GetForcedProhibitedPair()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::GetForcedProhibitedPair </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>constraintnumber</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const bool&#160;</td>
          <td class="paramname"><em>fiveprime</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Return a nucleotide in a prohibited pair. </p>
<p>This function returns either the five prime or three prime nucleotide in a prohibited pair constraint, depending on the value of fiveprime. Constraints are numbered from zero to GetNumberofForcedProhibited()-1. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">constraintnumber</td><td>is the index to the constraint number. </td></tr>
    <tr><td class="paramname">fiveprime</td><td>determines if the nucleotide is the five prime or the three prime nucleotide in the constraint. true = five prime nucleotide. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that is the nucleotide index. If the constraintnumber is for a constraint that does not exist, zero is returned. </dd></dl>

</div>
</div>
<a id="abe8372147747e249430353f7366c4310"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abe8372147747e249430353f7366c4310">&#9670;&nbsp;</a></span>GetForcedSingleStranded()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::GetForcedSingleStranded </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>constraintnumber</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Return a nucleotide that is forced single stranded. </p>
<p>This function returns a nucleotide that is constrainted to be single stranded. Constraints are numbered from zero to GetNumberofForcedSingleStranded()-1. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">constraintnumber</td><td>is the index to the constraint number. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that is the nucleotide index. If the constraintnumber is for a constraint that does not exist, zero is returned. </dd></dl>

</div>
</div>
<a id="a4534199d60cfffe08f0a89cd77f3acd9"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4534199d60cfffe08f0a89cd77f3acd9">&#9670;&nbsp;</a></span>GetFreeEnergy()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double RNA::GetFreeEnergy </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>structurenumber</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the folding free energy change for a predicted structure. </p>
<p>Returns the folding free energy change of structure i as determined by a previous folding calculation. Function requires that the structure be predicted by a structure prediction method. This function generates internal error codes that can be accessed by <a class="el" href="class_r_n_a.html#a08a3be93861700dce517caf7ccf77358" title="Return an error code, where a return of zero is no error. ">GetErrorCode()</a> after the constructor is called: 0 = no error, nonzero = error. The errorcode can be resolved to a c string using GetErrorMessage. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">structurenumber</td><td>is an integer indicating the predicted structure number. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A double that is the folding free energy change in kcal/mol. </dd></dl>

</div>
</div>
<a id="ab5201096c45cc26ec3a47c7cd98473a4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab5201096c45cc26ec3a47c7cd98473a4">&#9670;&nbsp;</a></span>GetFullErrorMessage()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">string RNA::GetFullErrorMessage </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>If there was an error, this returns the error message, along with error details (if any). if there was no error (i.e. <a class="el" href="class_r_n_a.html#a08a3be93861700dce517caf7ccf77358" title="Return an error code, where a return of zero is no error. ">GetErrorCode()</a> returns 0 and <a class="el" href="class_r_n_a.html#a29ba3acd2c1e8ed61e8e0c367cc6051c" title="Returns extended details about the last error. (e.g. error messages produced during file read operati...">GetErrorDetails()</a> returns NULL), this function returns an empty string (""); </p>

</div>
</div>
<a id="a23c879c487fcb254ebf645096026802e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a23c879c487fcb254ebf645096026802e">&#9670;&nbsp;</a></span>GetLabelXCoordinate()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::GetLabelXCoordinate </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>i</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the X coordinate for placing the nucleotide index label specified by i. </p>
<p>This function gets the X coordinate for placing the nucleotide index label specified by i. The user needs to have determined the coordinates for a complete structure using DetermineDrawingCoordinates prior to making this call. This function generates internal error codes that can be accessed by <a class="el" href="class_r_n_a.html#a08a3be93861700dce517caf7ccf77358" title="Return an error code, where a return of zero is no error. ">GetErrorCode()</a>: 0 = no error, nonzero = error. The errorcode can be resolved to a c string using GetErrorMessage. Zero is returned in case of error, but note that zero is also a valid coordinate. One additiona caveat: Labels that are placed at 0,0 are lables that would have overlapped nucleotides. These labels should not be drawn. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i</td><td>is an integer refering to the label to be drawn. This needs to be a multiple of 10. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An int that gives the X coordinate. </dd></dl>

</div>
</div>
<a id="affd3ef139bea53fbc39df41f68e1f030"></a>
<h2 class="memtitle"><span class="permalink"><a href="#affd3ef139bea53fbc39df41f68e1f030">&#9670;&nbsp;</a></span>GetLabelYCoordinate()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::GetLabelYCoordinate </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>i</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the Y coordinate for placing the nucleotide index label specified by i. </p>
<p>This function gets the Y coordinate for placing the nucleotide index label specified by i. The user needs to have determined the coordinates for a complete structure using DetermineDrawingCoordinates prior to making this call. This function generates internal error codes that can be accessed by <a class="el" href="class_r_n_a.html#a08a3be93861700dce517caf7ccf77358" title="Return an error code, where a return of zero is no error. ">GetErrorCode()</a>: 0 = no error, nonzero = error. The errorcode can be resolved to a c string using GetErrorMessage. Zero is returned in case of error, but note that zero is also a valid coordinate. One additiona caveat: Labels that are placed at 0,0 are lables that would have overlapped nucleotides. These labels should not be drawn. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i</td><td>is an integer refering to the label to be drawn. This needs to be a multiple of 10. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An int that gives the Y coordinate. </dd></dl>

</div>
</div>
<a id="a3bc6d5801335f85eed40a2daa987d96b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3bc6d5801335f85eed40a2daa987d96b">&#9670;&nbsp;</a></span>GetMaximumPairingDistance()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::GetMaximumPairingDistance </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Return the maximum pairing distance. </p>
<p>return An integer that indicates the maximum distance allowed between paired nucleotides, where -1 indicates that the maximum distance is not set. </p>

</div>
</div>
<a id="a84842b95ea62a357b12c440ad8ff4912"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a84842b95ea62a357b12c440ad8ff4912">&#9670;&nbsp;</a></span>GetNucleotide()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">char RNA::GetNucleotide </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>i</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>param An integer specifying the nucleotide index (starting at 1 and ending at <a class="el" href="class_r_n_a.html#a4ca21eb3e90dc475cc32c616b614ebc2" title="Get the total length of the sequence. ">GetSequenceLength()</a>). This function generates internal error codes that can be accessed by <a class="el" href="class_r_n_a.html#a08a3be93861700dce517caf7ccf77358" title="Return an error code, where a return of zero is no error. ">GetErrorCode()</a>: 0 = no error, nonzero = error. The errorcode can be resolved to a c string using GetErrorMessage. Note that nucleotides are numbered starting at an index of 1. return The char representing the nucleotide at index i or '-' if an error occured. </p>

</div>
</div>
<a id="a97c3e9a51d32a73fadd53f467c8f3ae0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a97c3e9a51d32a73fadd53f467c8f3ae0">&#9670;&nbsp;</a></span>GetNucleotideXCoordinate()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::GetNucleotideXCoordinate </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>i</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the X coordinate for nucleotide i for drawing a structure. </p>
<p>This function gets the X coordinate for placing the nucleotide specified by i. The user needs to have determined the coordinates for a complete structure using DetermineDrawingCoordinates prior to making this call. This function generates internal error codes that can be accessed by <a class="el" href="class_r_n_a.html#a08a3be93861700dce517caf7ccf77358" title="Return an error code, where a return of zero is no error. ">GetErrorCode()</a>: 0 = no error, nonzero = error. The errorcode can be resolved to a c string using GetErrorMessage. Zero is returned in case of error, but note that zero is also a valid coordinate. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i</td><td>is an integer refering to the nucleotide to be drawn. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An int that gives the X coordinate. </dd></dl>

</div>
</div>
<a id="acb2539f11b34a8e34668697ebf2890fb"></a>
<h2 class="memtitle"><span class="permalink"><a href="#acb2539f11b34a8e34668697ebf2890fb">&#9670;&nbsp;</a></span>GetNucleotideYCoordinate()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::GetNucleotideYCoordinate </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>i</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the Y coordinate for nucleotide i for drawing a structure. </p>
<p>This function gets the Y coordinate for placing the nucleotide specified by i. The user needs to have determined the coordinates for a complete structure using DetermineDrawingCoordinates prior to making this call. This function generates internal error codes that can be accessed by <a class="el" href="class_r_n_a.html#a08a3be93861700dce517caf7ccf77358" title="Return an error code, where a return of zero is no error. ">GetErrorCode()</a>: 0 = no error, nonzero = error. The errorcode can be resolved to a c string using GetErrorMessage. Zero is returned in case of error, but note that zero is also a valid coordinate. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i</td><td>is an integer refering to the nucleotide to be drawn. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An int that gives the Y coordinate. </dd></dl>

</div>
</div>
<a id="af9730b22a5500a3f23b4ee80382074e5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af9730b22a5500a3f23b4ee80382074e5">&#9670;&nbsp;</a></span>GetNumberOfForcedDoubleStranded()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::GetNumberOfForcedDoubleStranded </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Return the number of nucletides forced to be paired. </p>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates the number of nucleotides that are forced pair. </dd></dl>

</div>
</div>
<a id="a8a3843c206b1cf6790754454da102da4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8a3843c206b1cf6790754454da102da4">&#9670;&nbsp;</a></span>GetNumberOfForcedFMNCleavages()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::GetNumberOfForcedFMNCleavages </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Return the number of nucleotides accessible to FMN cleavage. </p>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates the number of FMN cleavage nucleotides (Us in GU pairs). </dd></dl>

</div>
</div>
<a id="a50c7d53681d5618f3c04e2140f8bb0c5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a50c7d53681d5618f3c04e2140f8bb0c5">&#9670;&nbsp;</a></span>GetNumberOfForcedModifications()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::GetNumberOfForcedModifications </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Return the number of nucleotides accessible to chemical modification. </p>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates the number of modified nucleotides. </dd></dl>

</div>
</div>
<a id="ae1c0ee1877897c83292d4194cc137206"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae1c0ee1877897c83292d4194cc137206">&#9670;&nbsp;</a></span>GetNumberOfForcedPairs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::GetNumberOfForcedPairs </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Return the number of forced base pairs. </p>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates the number of forced pairs. </dd></dl>

</div>
</div>
<a id="a11cd24997b6494ed631121b84e3b9b9b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a11cd24997b6494ed631121b84e3b9b9b">&#9670;&nbsp;</a></span>GetNumberOfForcedProhibitedPairs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::GetNumberOfForcedProhibitedPairs </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Return the number of prohibited base pairs. </p>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates the number of pairs that are prohibited. </dd></dl>

</div>
</div>
<a id="a505c57ee97ac3b5c1c21063087658ca3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a505c57ee97ac3b5c1c21063087658ca3">&#9670;&nbsp;</a></span>GetNumberOfForcedSingleStranded()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::GetNumberOfForcedSingleStranded </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Return the number of nucleotides that are not allowed to pair. </p>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates the number of nucleotides not allowed to pair. </dd></dl>

</div>
</div>
<a id="a7dfa253c1491b51383acd12b73c72a2f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7dfa253c1491b51383acd12b73c72a2f">&#9670;&nbsp;</a></span>GetPair()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::GetPair </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>i</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>structurenumber</em> = <code>1</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the nucleotide to which the specified nucleotide is paired. </p>
<p>Returns the pairing partner of the ith nucleotide in structure number structurenumber. Zero means the nucleotide is unpaired. This function generates internal error codes that can be accessed by <a class="el" href="class_r_n_a.html#a08a3be93861700dce517caf7ccf77358" title="Return an error code, where a return of zero is no error. ">GetErrorCode()</a> after the constructor is called: 0 = no error, nonzero = error. The errorcode can be resolved to a c string using GetErrorMessage. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i</td><td>is an int that indicates the nucleotide to which the pairing partner is being queried. </td></tr>
    <tr><td class="paramname">structurenumber</td><td>is an int that indicates the structure number, where the default is 1. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An int that indicates the other nucleotide in pair, where 0 is no paired. </dd></dl>

</div>
</div>
<a id="aa240d53bc9bc53897931a18b67bbcee0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa240d53bc9bc53897931a18b67bbcee0">&#9670;&nbsp;</a></span>GetPairEnergy()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double RNA::GetPairEnergy </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>i</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>j</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the lowest folding free energy possible for a structure containing pair i-j. </p>
<p>Returns a folding free energy change in kcal/mol for use in energy dot plots. This function requires that the <a class="el" href="class_r_n_a.html">RNA</a> constructor be called with a save file (.sav) name. (That is, for historical resaons, this cannot be called after FoldSingleStrand without writing the data to disk.) This function generates internal error codes that can be accessed by <a class="el" href="class_r_n_a.html#a08a3be93861700dce517caf7ccf77358" title="Return an error code, where a return of zero is no error. ">GetErrorCode()</a> after the constructor is called: 0 = no error, nonzero = error. The errorcode can be resolved to a c string using GetErrorMessage. param i and j are ints that provide indexes the 5' and 3' nucleotides, respectively, in a pair. return A double that is the folding free energy change in kcal/mol. </p>

</div>
</div>
<a id="a07e1fb189a45eb5b4604c286d5dd9d14"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a07e1fb189a45eb5b4604c286d5dd9d14">&#9670;&nbsp;</a></span>GetPairProbabilities()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::GetPairProbabilities </td>
          <td>(</td>
          <td class="paramtype">double *&#160;</td>
          <td class="paramname"><em>arr</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>size</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Fills an array with all basepair probabilities Function requires that the partition function data be present either because <a class="el" href="class_r_n_a.html#a654db191fe8a0a76350dfc57b2c5740c" title="Predict the partition function for a sequence. ">PartitionFunction()</a> has been called or the constructor that reads a partition function save was used. This function generates internal error codes that can be accessed by <a class="el" href="class_r_n_a.html#a08a3be93861700dce517caf7ccf77358" title="Return an error code, where a return of zero is no error. ">GetErrorCode()</a>: 0 = no error, nonzero = error. The errorcode can be resolved to a c string using GetErrorMessage. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">arr</td><td>A pointer to an array of doubles to receive the probabilities. The array should be allocated to hold (N-1)*N/2 elments, where N is the sequence length. The array will be filled with probabilities in upper-triangular format: e.g. for N=5 the array would be: P(1:2) P(1:3) P(1:4) P(1:5) P(2:3) P(2:4) P(2:5) P(3:4) P(3:5) P(4:5) (i.e. 5*4/2 = 10 elements) </td></tr>
    <tr><td class="paramname">size</td><td>The size of the array of doubles. If this is less than the required size, the array will not be modified and the function will return the required array size. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The total size required to hold the array of probabilities. This is equal to (N-1)*N/2 elments, where N is the sequence length. The function returns a negative number if an error occurred. Call GetErrorMessage(-returnValue) to get a textual description of the error. </dd></dl>

</div>
</div>
<a id="abd1da0a82663b38414a01c39b1a3973e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abd1da0a82663b38414a01c39b1a3973e">&#9670;&nbsp;</a></span>GetPairProbability()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double RNA::GetPairProbability </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>i</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>j</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get a base pair probability. </p>
<p>Returns the base pair probability for the pair between i and j. Function requires that the partition function data be present either because <a class="el" href="class_r_n_a.html#a654db191fe8a0a76350dfc57b2c5740c" title="Predict the partition function for a sequence. ">PartitionFunction()</a> has been called or the constructor that reads a partition function save was used. This function generates internal error codes that can be accessed by <a class="el" href="class_r_n_a.html#a08a3be93861700dce517caf7ccf77358" title="Return an error code, where a return of zero is no error. ">GetErrorCode()</a>: 0 = no error, nonzero = error. The errorcode can be resolved to a c string using GetErrorMessage. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i</td><td>provides the 5' nucleotide in a pair. </td></tr>
    <tr><td class="paramname">j</td><td>provides the 3' nucleotides in a pair. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A double that is the base pair probability. If i and j cannot pair, 0.0 is returned. If an error occurs, 0.0 is returned. </dd></dl>

</div>
</div>
<a id="a772f52c2865590b7a153ae3a6aab1de5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a772f52c2865590b7a153ae3a6aab1de5">&#9670;&nbsp;</a></span>GetProgress()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">ProgressHandler * RNA::GetProgress </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Return the current pointer to an instance of a class derived from ProgressHandler (such as TProgressDialog for text-mode programs or a simple ProgressHandler for GUI progams or scripts) This is used during inheritance to provide access to the underlying TProgressDialog. </p>

</div>
</div>
<a id="ac3b16f4e3c72091bf69a55e923c51728"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac3b16f4e3c72091bf69a55e923c51728">&#9670;&nbsp;</a></span>GetSequence() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">const char * RNA::GetSequence </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<dl class="section return"><dt>Returns</dt><dd>The full nucleotide sequence. </dd></dl>

</div>
</div>
<a id="a9f49a0851278aaa5aa39a1d538c58fa9"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9f49a0851278aaa5aa39a1d538c58fa9">&#9670;&nbsp;</a></span>GetSequence() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::string RNA::GetSequence </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>length</em> = <code>std::string::npos</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<dl class="section return"><dt>Returns</dt><dd>A segment of the nucleotide sequence. </dd></dl>

</div>
</div>
<a id="a4ca21eb3e90dc475cc32c616b614ebc2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4ca21eb3e90dc475cc32c616b614ebc2">&#9670;&nbsp;</a></span>GetSequenceLength()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::GetSequenceLength </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the total length of the sequence. </p>
<dl class="section return"><dt>Returns</dt><dd>An integer that specifies the total length of the sequence. </dd></dl>

</div>
</div>
<a id="a7513721374fab949fd42bbe4980894dc"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7513721374fab949fd42bbe4980894dc">&#9670;&nbsp;</a></span>GetStructure()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">structure * RNA::GetStructure </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Access the underlying structure class. This is provided for use with two sequence methods. Generally, there is no need for end users to use this function because the <a class="el" href="class_r_n_a.html">RNA</a> class provides an convenient wrapper for accessing the information in an <a class="el" href="class_r_n_a.html">RNA</a> class. </p><dl class="section return"><dt>Returns</dt><dd>A pointer to structure. </dd></dl>

</div>
</div>
<a id="a27b3bdb17fb4233cc4f91753beda7570"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a27b3bdb17fb4233cc4f91753beda7570">&#9670;&nbsp;</a></span>GetStructureNumber()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::GetStructureNumber </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the total number of specified or predicted structures. </p>
<dl class="section return"><dt>Returns</dt><dd>An integer specify the total number of structures. </dd></dl>

</div>
</div>
<a id="a78bdc62dac46d0d817fed67f5843010c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a78bdc62dac46d0d817fed67f5843010c">&#9670;&nbsp;</a></span>GetVprimeQ()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double RNA::GetVprimeQ </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>i</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>j</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="ad1d51357d72c2be6b5f62c31594d401b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad1d51357d72c2be6b5f62c31594d401b">&#9670;&nbsp;</a></span>GetW()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double RNA::GetW </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>i</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>j</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a4eb047c1e53a52c0ce515e0289e6e84f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4eb047c1e53a52c0ce515e0289e6e84f">&#9670;&nbsp;</a></span>init()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void RNA::init </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>sequenceOrFileName</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="_r_n_a_8h.html#a3883db120ec3ca0fe72f322a76dd2378">RNAInputType</a>&#160;</td>
          <td class="paramname"><em>fileType</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const bool&#160;</td>
          <td class="paramname"><em>allowUnknownBases</em> = <code>false</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const bool&#160;</td>
          <td class="paramname"><em>skipThermoTables</em> = <code>false</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="a61141a694126843efe07081e5918fba7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a61141a694126843efe07081e5918fba7">&#9670;&nbsp;</a></span>MaximizeExpectedAccuracy()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::MaximizeExpectedAccuracy </td>
          <td>(</td>
          <td class="paramtype">const double&#160;</td>
          <td class="paramname"><em>maxPercent</em> = <code>20</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>maxStructures</em> = <code>20</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>window</em> = <code>1</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const double&#160;</td>
          <td class="paramname"><em>gamma</em> = <code>1.0</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Predict the structure with maximum expected accuracy and suboptimal structures. </p>
<p>This function predicts structures composed of probable base pairs and single-srtranded nucleotide, weighted by gamma. The score for a structure is = gamma * 2 * (sum of pairing probabilities for pairs) + (sum of unpairing probabilities for single stranded nucleotides). This function requires partition function data from either a previous partition function calculations or from having read a partition function save file during construction of the class. In case of error, the function returns a non-zero that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">maxPercent</td><td>is the maximum percent difference is score in generating suboptimal structures. The default is 20. </td></tr>
    <tr><td class="paramname">maxStructures</td><td>is the maximum number of suboptimal structures to generate. The default is 20. </td></tr>
    <tr><td class="paramname">window</td><td>is the window parameter, where a higher value generates suboptimal structures that are more different from each other. The default is 1. </td></tr>
    <tr><td class="paramname">gamma</td><td>is the weight given to base pairs </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An int that indicates an error code (0 = no error, non-zero = error). </dd></dl>

</div>
</div>
<a id="a654db191fe8a0a76350dfc57b2c5740c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a654db191fe8a0a76350dfc57b2c5740c">&#9670;&nbsp;</a></span>PartitionFunction()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::PartitionFunction </td>
          <td>(</td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>savefile</em>[] = <code>&quot;&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>temperature</em> = <code>-10.0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>disablecoax</em> = <code>false</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>restoreSHAPE</em> = <code>true</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>allowisolated</em> = <code>false</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>maxinter</em> = <code>30</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Predict the partition function for a sequence. </p>
<p>This function must be called to predict base pair probabilities, perform stochastic traceback, or for maximizing expected accuracy. If the temperature has not been specified using SetTemperature and no free energies have been calculated, the thermodynamic parameters have not been read and therefore they will be read by this function call. The parameter files should be located in the directory specified by the environment variable $DATAPATH of the pwd. In case of error, the function returns a non-zero that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. Note that the parameter temperature is used when calculating equilibrium constants, but does not change the temperature at which the free energies are determined. SetTemperature, from the underlying base class <a class="el" href="class_thermodynamics.html" title="Thermodynamics Class. ">Thermodynamics</a>, should be used to change the temperature for most calculations. This parameter should generally not be used. The default is -10.0 and values below zero cause this parameter to be ignored (the correct default behavior). Note also that if SetTemperature is not used, the temperature defaults to 310.15 K (37 deg. C), which is the desired behavior for most purposes. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">savefile</td><td>is a c string that contains the path and filename for creating a save file. This defaults to "", which indicates no file is to be written. </td></tr>
    <tr><td class="paramname">temperature</td><td>is a double that indicates a pseudo-temperature for calculating equilibrium constants from free energies at fixed temperature previously specified. </td></tr>
    <tr><td class="paramname">allowisolated</td><td>is a bool that indicates if isolated base pairs should be allowed. The default, false, uses a heuristic to filter out most isolated base pairs. </td></tr>
    <tr><td class="paramname">restoreSHAPE</td><td>Whether SHAPE data should be restored after this call. The default is true. PartitionFunction converts SHAPE pseudo-energies into equillibrium constants, overwriting the <a class="el" href="class_r_n_a.html#a7513721374fab949fd42bbe4980894dc">GetStructure()</a>-&gt;SHAPE array. If restoreSHAPE is true, the original values will be restored before the function returns. Otherwise the equillibrium constants will remain in the <a class="el" href="class_r_n_a.html#a7513721374fab949fd42bbe4980894dc">GetStructure()</a>-&gt;SHAPE array. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An int that indicates an error code (0 = no error, 5 = error reading thermodynamic parameter files). </dd></dl>

</div>
</div>
<a id="ad05388cc60648f09cc1acb3c4b513516"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad05388cc60648f09cc1acb3c4b513516">&#9670;&nbsp;</a></span>PredictProbablePairs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::PredictProbablePairs </td>
          <td>(</td>
          <td class="paramtype">const float&#160;</td>
          <td class="paramname"><em>probability</em> = <code>0</code></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Predict structures containing highly probable pairs. </p>
<p>This function predicts structures composed of probable base pairs. This function requires partition function data from either a previous partition function calculations or from having read a partition function save file during construction of the class. In case of error, the function returns a non-zero that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">probability</td><td>is the pairing probability threshold, where pairs will be predicted if they have a higher probability. Note that a value of less than 0.5 (50%), will cause an error. The default value of zero will trigger the creation of 8 structures, with thresholds of &gt;=0.99, &gt;=0.97, &gt;=0.95, &gt;=0.90, &gt;=0.80, &gt;=0.70, &gt;=0.60, &gt;0.50. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An int that indicates an error code (0 = no error, non-zero = error). </dd></dl>

</div>
</div>
<a id="ab5dd63132dc8324c84c670cf887884c2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab5dd63132dc8324c84c670cf887884c2">&#9670;&nbsp;</a></span>ProbKnot()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::ProbKnot </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>iterations</em> = <code>1</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>MinHelixLength</em> = <code>1</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>threshold</em> = <code>0</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Predict maximum expected accuracy structures that contain pseudoknots from either a sequence or a partition function save file. </p>
<p>This function uses base pair probabilities to predict structures that contains pseudoknots. This function requires partition function data from either a previous partition function calculations or from having read a partition function save file during construction of the class. In case of error, the function returns a non-zero that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">iterations</td><td>is the number of iterations of pair selection that are performed. The default and recommended value is 1. </td></tr>
    <tr><td class="paramname">MinHelixLength</td><td>is the shortest helix that is allowed. If this is set &gt;1, a post-processing step is performed to remove short helices. Default = 1, i.e. no post-processing. </td></tr>
    <tr><td class="paramname">threshold</td><td>is the required minimum probability for a pair. Only include pairs with a higher probability. Default = 0, i.e. include all pairs. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An int that indicates an error code (0 = no error, non-zero = error). </dd></dl>

</div>
</div>
<a id="a2049ad01ea76e5072133e23dbd24d03d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2049ad01ea76e5072133e23dbd24d03d">&#9670;&nbsp;</a></span>ProbKnotFromSample()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::ProbKnotFromSample </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>iterations</em> = <code>1</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>MinHelixLength</em> = <code>1</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>threshold</em> = <code>0</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Predict maximum expected accuracy structures that contain pseudoknots from a file containing ensemble of structures. </p>
<p>This function uses base pair probabilities to predict structures that contains pseudoknots. This function requires a file with ensemble of structures. This function processes the file to calculate pair probabilities. In case of error, the function returns a non-zero that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">iterations</td><td>is the number of iterations of pair selection that are performed. The default and recommended value is 1. </td></tr>
    <tr><td class="paramname">MinHelixLength</td><td>is the shortest helix that is allowed. If this is set &gt;1, a post-processing step is performed to remove short helices. Default = 1, i.e. no post-processing. </td></tr>
    <tr><td class="paramname">threshold</td><td>is the required minimum probability for a pair. Only include pairs with a higher probability. Default = 0, i.e. include all pairs. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An int that indicates an error code (0 = no error, non-zero = error). </dd></dl>

</div>
</div>
<a id="a256ed0d8b5a7fdd1dc1ea5ced058cc7a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a256ed0d8b5a7fdd1dc1ea5ced058cc7a">&#9670;&nbsp;</a></span>ReadConstraints()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::ReadConstraints </td>
          <td>(</td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>filename</em>[]</td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Read a set of folding constraints to disk in a plain text file. </p>
<p>The file format for constraints is that generated by the <a class="el" href="class_r_n_a.html#aa860f3f448840a8af8461212b1920ab6" title="Write the current set of folding constraints to disk in a plain text file. ">WriteConstraints()</a> function. The function returns 0 with no error and a non-zero otherwise that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. Note that calling <a class="el" href="class_r_n_a.html#a256ed0d8b5a7fdd1dc1ea5ced058cc7a" title="Read a set of folding constraints to disk in a plain text file. ">ReadConstraints()</a> will erase previously defined constraints (except for SHAPE pseudoenergy restraints). </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">filename</td><td>is a c string that is the file name to be read. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates an error code (0 = no error, 1 = file not found, 13 = error reading constraint file). </dd></dl>

</div>
</div>
<a id="a0a8bba2d4d6c3a9afd718a7bf420eb29"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a0a8bba2d4d6c3a9afd718a7bf420eb29">&#9670;&nbsp;</a></span>ReadDMS()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::ReadDMS </td>
          <td>(</td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>filename</em>[], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const bool&#160;</td>
          <td class="paramname"><em>bynt</em> = <code>false</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Read DMS data to constrain structure prediction on subsequent structure predictions. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">filename</td><td>is a c string that indicates a file that contains DMS data. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Returns an integer that indicates an error code (0 = no error, 1 = input file not found). </dd></dl>

</div>
</div>
<a id="a05b4ec1bf9e25c5673c065baabab6233"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a05b4ec1bf9e25c5673c065baabab6233">&#9670;&nbsp;</a></span>ReadDSO()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::ReadDSO </td>
          <td>(</td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>filename</em>[]</td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Read double strand offset data from disk. </p>
<p>The double strand offset is data that is used to constrain structure prediction on subsequent structure predictions. This is a free energy in kcal/mol that is added to a specific nucleotide that is double stranded. The function returns 0 with no error and a non-zero otherwise that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">filename</td><td>is a c string that indicates a file that contains data, in a raw format with nucleotide index and offset (one set per line). </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates an error code (0 = no error, 1 = input file not found). </dd></dl>

</div>
</div>
<a id="ad15624f880c955d1985914292b95d746"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad15624f880c955d1985914292b95d746">&#9670;&nbsp;</a></span>ReadExperimentalPairBonus()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::ReadExperimentalPairBonus </td>
          <td>(</td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>filename</em>[], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double const&#160;</td>
          <td class="paramname"><em>experimentalOffset</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double const&#160;</td>
          <td class="paramname"><em>experimentalScaling</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Read experimental pair bonuses from disk. </p>
<p>This is a quantity that results in a bonus added to a specific pair, once per stack, so that pairs in the middle of a helix get the bonus twice and those at the end of a helix get the bonus once. The bonus is in the form of experimentalScaling*value + experimentalOffset. The data is formatted using a simple square matrix of values and no headers. The format requires that there be N^2 entries for a sequence of N nucleotides. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">filename</td><td>is a c string that indicates a file that contains data. </td></tr>
    <tr><td class="paramname">experimentalOffset</td><td>is a double that is added to each value. </td></tr>
    <tr><td class="paramname">experimentalScaling</td><td>is a double by which each value is multiplied. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates an error code (0 = no error, 1 = input file not found). </dd></dl>

</div>
</div>
<a id="a54e154d693ade8fcfe62e47961d37701"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a54e154d693ade8fcfe62e47961d37701">&#9670;&nbsp;</a></span>ReadSHAPE() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::ReadSHAPE </td>
          <td>(</td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>filename</em>[], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const double&#160;</td>
          <td class="paramname"><em>slope</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const double&#160;</td>
          <td class="paramname"><em>intercept</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">RestraintType&#160;</td>
          <td class="paramname"><em>modifier</em> = <code>RESTRAINT_SHAPE</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const bool&#160;</td>
          <td class="paramname"><em>IsPseudoEnergy</em> = <code>true</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Read SHAPE data from disk. </p>
<p>The SHAPE data is used to constrain structure prediction on subsequent structure predictions. The function returns 0 with no error and a non-zero otherwise that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. Pseudo folding free energy change parameters should be in units of kcal/mol. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">filename</td><td>is a c string that indicates a file that contains SHAPE data. </td></tr>
    <tr><td class="paramname">IsPseudoEnergy</td><td>indicates whether this is the pseudo folding free energy constraint (the preferred method). This defaults to true. </td></tr>
    <tr><td class="paramname">slope</td><td>is the slope when IsPseudoEnergy=true and is a threshold above which nucleotides are forced single stranded otherwise. </td></tr>
    <tr><td class="paramname">intercept</td><td>is the intercept when IsPseudoEnergy=true and is a threshold above which a nucleotide is considered chemically modified otherwise. </td></tr>
    <tr><td class="paramname">modifier</td><td>is the type of chemical modification probe that was used (currently accepted values are SHAPE, diffSHAPE, DMS, and CMCT). Defaults to SHAPE. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates an error code (0 = no error, 1 = input file not found). </dd></dl>

</div>
</div>
<a id="adb84ef498e9d26d3f7add0e47855621a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#adb84ef498e9d26d3f7add0e47855621a">&#9670;&nbsp;</a></span>ReadSHAPE() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::ReadSHAPE </td>
          <td>(</td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>filename</em>[], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const double&#160;</td>
          <td class="paramname"><em>dsSlope</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const double&#160;</td>
          <td class="paramname"><em>dsIntercept</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const double&#160;</td>
          <td class="paramname"><em>ssSlope</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const double&#160;</td>
          <td class="paramname"><em>ssIntercept</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">RestraintType&#160;</td>
          <td class="paramname"><em>modifier</em> = <code>RESTRAINT_SHAPE</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Read SHAPE data from disk including single-stranded SHAPE pseudo free energys. </p>
<p>The SHAPE data is used to constrain structure prediction on subsequent structure predictions. This version of the overloaded function includes a single-stranded pseudo free energy change. The function returns 0 with no error and a non-zero otherwise that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. Pseudo folding free energy change parameters should be in units of kcal/mol. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">filename</td><td>is a c string that indicates a file that contains SHAPE data. </td></tr>
    <tr><td class="paramname">dsSlope</td><td>is the double-stranded slope. </td></tr>
    <tr><td class="paramname">dsIntercept</td><td>is the double-stranded intercept. </td></tr>
    <tr><td class="paramname">modifier</td><td>is the type of chemical modification probe that was used (currently accepted values are SHAPE, DMS, and CMCT). Defaults to SHAPE. </td></tr>
    <tr><td class="paramname">ssSlope</td><td>is the single-stranded slope. </td></tr>
    <tr><td class="paramname">ssIntercept</td><td>is the single-stranded intercept. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates an error code (0 = no error, 1 = input file not found). </dd></dl>

</div>
</div>
<a id="a6ed8d145eb78bfcb22aa5f5e7e41f7c8"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6ed8d145eb78bfcb22aa5f5e7e41f7c8">&#9670;&nbsp;</a></span>ReadSSO()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::ReadSSO </td>
          <td>(</td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>filename</em>[]</td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Read single strand offset data from disk. </p>
<p>The single strand offset is data that is used to constrain structure prediction on subsequent structure predictions. This is a free energy in kcal/mol that is added to a specific nucleotide that is single stranded. The function returns 0 with no error and a non-zero otherwise that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">filename</td><td>is a c string that indicates a file that contains data, in a raw format with nucleotide index and offset (one set per line). </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates an error code (0 = no error, 1 = input file not found). </dd></dl>

</div>
</div>
<a id="ac068513a05333629af4fad23a5b3213d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac068513a05333629af4fad23a5b3213d">&#9670;&nbsp;</a></span>ReFoldSingleStrand()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::ReFoldSingleStrand </td>
          <td>(</td>
          <td class="paramtype">const float&#160;</td>
          <td class="paramname"><em>percent</em> = <code>20</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>maximumstructures</em> = <code>20</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>window</em> = <code>5</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Re-predict the lowest free energy secondary structure and generate suboptimal structures using a heuristic. </p>
<p>This function predicts the lowest free energy structure and suboptimal structure after a save file (.sav) was specified to the constructor. The step of predicting structures from the save file is rapid, so this is laregely a method to quickly generate a different set of suboptimal structures. Refolding can only be performed if the <a class="el" href="class_r_n_a.html">RNA</a> constructor was called with a save file name. (That is, you cannot call this after calling fold single strand, without loading the data from disk with a new instance of <a class="el" href="class_r_n_a.html">RNA</a>. This is for historical reasons.) In case of error, the function returns a non-zero that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">percent</td><td>is the maximum % difference in free energy in suboptimal structures from the lowest free energy structure. The default is 20. </td></tr>
    <tr><td class="paramname">maximumstructures</td><td>is the maximum number of suboptimal structures to generate. The default is 20. </td></tr>
    <tr><td class="paramname">window</td><td>is a parameter that specifies how different the suboptimal structures should be from each other (0=no restriction and larger integers require structures to be more different). The default is 5. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An int that indicates an error code (0 = no error, 5 = error reading thermodynamic parameter files, 14 = traceback error). </dd></dl>

</div>
</div>
<a id="a7aae4a7dcb9fcd87d3cf0c4c2393b9b3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7aae4a7dcb9fcd87d3cf0c4c2393b9b3">&#9670;&nbsp;</a></span>RemoveBasePair()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::RemoveBasePair </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>i</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>structurenumber</em> = <code>1</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Remove a specified pair in a specified structure. </p>
<p>Break the pair between i and i's pairing partner Return 0 if there is no error. Return 3 if structurenumber out of range. Return 4 if nucleotide number out of range. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i</td><td>is the index of a nucleotide in a pair that will be broken. </td></tr>
    <tr><td class="paramname">structurenumber</td><td>is an integer specifying the structure from which to remove the pairs. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates an error code that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>, 0 = no error. </dd></dl>

</div>
</div>
<a id="a3b4a3b4f525013b59f48b63423108d3c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3b4a3b4f525013b59f48b63423108d3c">&#9670;&nbsp;</a></span>RemoveConstraints()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void RNA::RemoveConstraints </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Remove all folding constraints. </p>
<p>This function strips all previously assigned folding constraints. Note that this function does not delete SHAPE constraints or pseudo free energies. </p>

</div>
</div>
<a id="ab6f64d9425687dc03f7cedf417bc9bc3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab6f64d9425687dc03f7cedf417bc9bc3">&#9670;&nbsp;</a></span>RemovePairs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::RemovePairs </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>structurenumber</em> = <code>1</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>removeIfLastStructure</em> = <code>true</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Remove all the current base pairs in a specified structure. </p>
<p>Return 0 if there is no error. Return 3 if there are no structures or if structurenumber is out of range. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">structurenumber</td><td>is an integer specifying the structure from which to remove the pairs. </td></tr>
    <tr><td class="paramname">removeIfLastStructure</td><td>If true and this is the last structure in the group, it will be removed. (default: true) </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates an error code that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>, 0 = no error. </dd></dl>

</div>
</div>
<a id="a5e8cc3e54438c3a4ed2a042723386006"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5e8cc3e54438c3a4ed2a042723386006">&#9670;&nbsp;</a></span>ResetError()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void RNA::ResetError </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Reset the <a class="el" href="class_r_n_a.html">RNA</a>'s internal error code to 0 This should be invoked after the error condition is handled </p>

</div>
</div>
<a id="aac5de92b17d1fe4d2eb0cc3fbee595aa"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aac5de92b17d1fe4d2eb0cc3fbee595aa">&#9670;&nbsp;</a></span>Rsample()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::Rsample </td>
          <td>(</td>
          <td class="paramtype">const vector&lt; double &gt; &amp;&#160;</td>
          <td class="paramname"><em>experimentalRestraints</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">RsampleData &amp;&#160;</td>
          <td class="paramname"><em>refdata</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>randomSeed</em> = <code>0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>savefile</em>[] = <code>&quot;&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const double&#160;</td>
          <td class="paramname"><em>cparam</em> = <code>0.5</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const double&#160;</td>
          <td class="paramname"><em>offset</em> = <code>1.10</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>numsamples</em> = <code>10000</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Rsample function calculates partition function with the pseudo-dg restraints from Rsample algorithm. It calls partition function twice. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">experimentalRestraints</td><td>is a 0-based vector of experimental restraints (such as SHAPE or DMS). </td></tr>
    <tr><td class="paramname">refdata</td><td>are 3 vectors read by RsampleData calss that contain paired-end, paired-middle and unpaired reference reactivities. </td></tr>
    <tr><td class="paramname">savefile</td><td>is a c string that contains the path and filename for creating a save file. This defaults to "", which indicates no file is to be written. </td></tr>
    <tr><td class="paramname">cparam</td><td>is the C parameter that Rsample algorith uses. Default value is 0.5. </td></tr>
    <tr><td class="paramname">offset</td><td>is the Offset parameter that Rsample algorith uses. Default value is 1.1. </td></tr>
    <tr><td class="paramname">numsamples</td><td>is the number of samples to get from stochastic sampling routine called inside Rsample function. Default is 10,000. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An int that indicates an error code (0 = no error). </dd></dl>

</div>
</div>
<a id="ac7a793cd3862b976a5b354da43a60df2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac7a793cd3862b976a5b354da43a60df2">&#9670;&nbsp;</a></span>SetConstraints()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void RNA::SetConstraints </td>
          <td>(</td>
          <td class="paramtype">vector&lt; int &gt;&#160;</td>
          <td class="paramname"><em>ss</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="ae45785f80ecd7b6b7783fb290dd78152"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae45785f80ecd7b6b7783fb290dd78152">&#9670;&nbsp;</a></span>SetErrorDetails()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void RNA::SetErrorDetails </td>
          <td>(</td>
          <td class="paramtype">const string &amp;&#160;</td>
          <td class="paramname"><em>details</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Set extended details about the last error. (e.g. error messages produced during file read operations that are otherwise lost.) </p>

</div>
</div>
<a id="a127218155fa5e2d68aaf6729e97616e0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a127218155fa5e2d68aaf6729e97616e0">&#9670;&nbsp;</a></span>SetExtrinsic()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::SetExtrinsic </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>i</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>j</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>k</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Add extrinsic restraints for partition function calculations. </p>
<p>This function multiplies the equilibrium constant for structures including the i-j basepair by k. This applies only to partition functions and to stochastic traceback. If k&gt;1, then the i-j pair is favored and if k&lt;1, the i-j pair is disfavored. k should always be &gt;= 0. In case of error, the function returns a non-zero that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i</td><td>is the index of a nucleotide in the i-j pair. </td></tr>
    <tr><td class="paramname">j</td><td>is the index of the other nucleotide in the i-j pair. </td></tr>
    <tr><td class="paramname">k</td><td>is an equilibrium constant that is &gt;= 0. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates an error code (0 = no error, &gt;0 indicates an error). </dd></dl>

</div>
</div>
<a id="a507cd74f3b8a98e2172ec7c85465d919"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a507cd74f3b8a98e2172ec7c85465d919">&#9670;&nbsp;</a></span>SetProgress()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void RNA::SetProgress </td>
          <td>(</td>
          <td class="paramtype">ProgressHandler &amp;&#160;</td>
          <td class="paramname"><em>Progress</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Provide a reference to an instance of a class derived from ProgressHandler (such as TProgressDialog for text-mode programs or a simple ProgressHandler for GUI progams or scripts) A ProgressHandler class has a public function void update(int percent) that indicates the progress of a long calculation. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">Progress</td><td>is an instance of a ProgressHandler, TProgressDialog, or PartialProgress etc. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="aa746e33612c6cc7fcf8856fac3d364ac"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa746e33612c6cc7fcf8856fac3d364ac">&#9670;&nbsp;</a></span>SetSequenceLabel()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void RNA::SetSequenceLabel </td>
          <td>(</td>
          <td class="paramtype">const string &amp;&#160;</td>
          <td class="paramname"><em>label</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Set the name of the sequence. (This must be called before structure prediction for it to affect the resulting structures.) </p>

</div>
</div>
<a id="ac529346af902b21f9b3bb2b17a3ee661"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac529346af902b21f9b3bb2b17a3ee661">&#9670;&nbsp;</a></span>SpecifyPair()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::SpecifyPair </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>i</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>j</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>structurenumber</em> = <code>1</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Specify a base pair between nucleotides i and j. </p>
<p>The base pair is in structure number structurenumber, which is assumed to be structure 1. Return 0 if there is no problem, otherwise return an error code: error = 3 -&gt; structurenumber out of range. error = 4 -&gt; nucleotide number out of range. A c string or string description of the error are available using <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. Note!: Sequences with the 5' end = nucleotide 1. Note!: Structures start at structure 1. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i</td><td>is an integer for the position of the first nucleotide in the pair. </td></tr>
    <tr><td class="paramname">j</td><td>in an integer for the position of the second nucleotide in the pair. </td></tr>
    <tr><td class="paramname">structurenumber</td><td>is the structure that has the pair. This defaults to 1. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates an error code that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>, 0 = no error. </dd></dl>

</div>
</div>
<a id="ab06ab0d0af7f74d3e3a20c8d52e65ef5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab06ab0d0af7f74d3e3a20c8d52e65ef5">&#9670;&nbsp;</a></span>Stochastic()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::Stochastic </td>
          <td>(</td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>structures</em> = <code>1000</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>seed</em> = <code>1</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Sample structures from the Boltzman ensemable. </p>
<p>This function requires partition function data from either a previous partition function calculations or from having read a partition function save file during construction of the class. In case of error, the function returns a non-zero that can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">structures</td><td>is the number of structures to be sampled. The default is 1000. </td></tr>
    <tr><td class="paramname">seed</td><td>is an integer that seeds the random number generator that is required for sampling, which defaults to 1. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An int that indicates an error code (0 = no error, non-zero = error). </dd></dl>

</div>
</div>
<a id="ae59b36ca61c62813ad096eebb02f3850"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae59b36ca61c62813ad096eebb02f3850">&#9670;&nbsp;</a></span>StopProgress()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void RNA::StopProgress </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Provide a means to stop using a TProgressDialog. StopProgress tells the <a class="el" href="class_r_n_a.html">RNA</a> class to no longer follow progress. This should be called if the TProgressDialog is deleted, so that this class does not make reference to it. </p>

</div>
</div>
<a id="aa860f3f448840a8af8461212b1920ab6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa860f3f448840a8af8461212b1920ab6">&#9670;&nbsp;</a></span>WriteConstraints()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::WriteConstraints </td>
          <td>(</td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>filename</em>[]</td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Write the current set of folding constraints to disk in a plain text file. </p>
<p>This function does not write SHAPE pseudo energies. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">filename</td><td>is a c string that is the file name to be written. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that indicates an error code (0 = no error). Currently, this function does not generate errors, but the return is provided to add error handling in the future. </dd></dl>

</div>
</div>
<a id="a6e1e0db7f8f0f3d9cb7dc14f5dc61953"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6e1e0db7f8f0f3d9cb7dc14f5dc61953">&#9670;&nbsp;</a></span>WriteCt()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::WriteCt </td>
          <td>(</td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>filename</em>[], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>append</em> = <code>false</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">CTCommentProvider &amp;&#160;</td>
          <td class="paramname"><em>commentProvider</em> = <code>CTComments::Energy</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Write a ct file of the structures. </p>
<p>Return 0 if no error and non-zero errors can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">filename</td><td>is a NULL terminated c string that specifies the name of the ct file to be written. </td></tr>
    <tr><td class="paramname">append</td><td>is a bool that indiactes whether the ct data should be appended to an existing file. If true, data will be appended if the file exists, or a new file will be created if the file does not exist. If false (the default), any existing file is overwritten. </td></tr>
    <tr><td class="paramname">commentProvider</td><td>a reference to an object that returns application-defined comments for each structure in a CT file. (e.g. "ENERGY = ..." ) This parameter allows a program to customize when comments are written and how they appear. For example many programs write "ENERGY = ..." but only when GetEnergy() != 0. However, a program like MaxExpect might write "SCORE = ..." (even if GetEnergy() returns zero). Similarly, Design might want to add comments like "NED = 0.02" which would require a different source of information than GetEnergy() &ndash; or at the very least, different formatting/precision. To provide custom comments, pass an in an object derived from CTCommentProvider that has overridden the getComment function. To disable comments, pass in CTComments::None. The default is CTComments::Energy, which provides legacy "ENERGY = ..." comments. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that provides an error code. 0 = no error. Use GetErrorMessage to get a textual description of the error. </dd></dl>

</div>
</div>
<a id="a2c04e45d3cf6436afe8622d140f2f1d1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2c04e45d3cf6436afe8622d140f2f1d1">&#9670;&nbsp;</a></span>WriteDotBracket()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::WriteDotBracket </td>
          <td>(</td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>filename</em>[], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>structurenumber</em> = <code>-1</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const DotBracketFormat&#160;</td>
          <td class="paramname"><em>format</em> = <code>DBN_FMT_MULTI_TITLE</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">CTCommentProvider &amp;&#160;</td>
          <td class="paramname"><em>commentProvider</em> = <code>CTComments::Energy</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Write dot-bracket file of structures. </p>
<p>Return 0 if no error and non-zero errors can be parsed by <a class="el" href="class_r_n_a.html#a939a33d1f65d21c257bbb2a0a2505953" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessage()</a> or <a class="el" href="class_r_n_a.html#a3e1f322c241c729f36b56cb0e452c0c4" title="Return error messages based on code from GetErrorCode and other error codes. ">GetErrorMessageString()</a>. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">filename</td><td>is a NULL terminated c string that specified the name of the file to be written. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An integer that provides an error code. 0 = no error. </dd></dl>

</div>
</div>
<a id="ab08f9c44b784e65ab966583e420fcba4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab08f9c44b784e65ab966583e420fcba4">&#9670;&nbsp;</a></span>WriteThermodynamicDetails()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::WriteThermodynamicDetails </td>
          <td>(</td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>filename</em>[], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const bool&#160;</td>
          <td class="paramname"><em>UseSimpleMBLoopRules</em> = <code>false</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Calculate the folding free energy change for all structures and write the details of the calculation to a file. </p>
<p>Free energies are in kcal/mol. The first time this is called, if no other free energy calculation has been performed and the folding temperature has not been specifed, thermodynamic parameter files (.dat) files will be read from disk. The parameter files should be located in the directory specified by environment variable $DATAPATH, or the pwd. In case of error, the function returns a non-zero. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">filename</td><td>is a c-string that provides the name of the output file to be written. If filename==NULL, free energies will be calculated, but no file will be written. </td></tr>
    <tr><td class="paramname">UseSimpleMBLoopRules</td><td>is a bool that indicates what energy rules to use. The default, false, uses the complete nearest neighbor model for multibranch loops. When true is passed, the energy model is instead a simplified model that is the one used by the dynamic programming algorithms. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An int that indicates whether an error occurred (0 = no error; 5 = error reading parameter files). </dd></dl>

</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a id="a408ad69397beb1ffd831191114fd7630"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a408ad69397beb1ffd831191114fd7630">&#9670;&nbsp;</a></span>ct</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">structure* RNA::ct</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="ae25e6cb1b35e47ab74890167fdc2cac4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae25e6cb1b35e47ab74890167fdc2cac4">&#9670;&nbsp;</a></span>drawallocated</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">bool RNA::drawallocated</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="a3c34e05e63dc1941784f1108b119019e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3c34e05e63dc1941784f1108b119019e">&#9670;&nbsp;</a></span>energyallocated</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">bool RNA::energyallocated</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="acf869e72c097bae7824a72e8771e7a5a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#acf869e72c097bae7824a72e8771e7a5a">&#9670;&nbsp;</a></span>ErrorCode</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::ErrorCode</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="ace814d16ca5fedb855c9aa0f6f2bc9b4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ace814d16ca5fedb855c9aa0f6f2bc9b4">&#9670;&nbsp;</a></span>ev</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">DynProgArray&lt;integersize&gt;* RNA::ev</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="ae265114cc811d410c0ea14a20a2fdb0b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae265114cc811d410c0ea14a20a2fdb0b">&#9670;&nbsp;</a></span>ew</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">DynProgArray&lt;integersize&gt; * RNA::ew</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="a0df9aab831d67cd88c2c0c95daef95bf"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a0df9aab831d67cd88c2c0c95daef95bf">&#9670;&nbsp;</a></span>ew2</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">DynProgArray&lt;integersize&gt;* RNA::ew2</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="a34ff98ed1d5627cfaed571447a4a71fb"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a34ff98ed1d5627cfaed571447a4a71fb">&#9670;&nbsp;</a></span>ew3</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">integersize * RNA::ew3</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="a43e0afefc5122174ea5f3065e58d0a29"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a43e0afefc5122174ea5f3065e58d0a29">&#9670;&nbsp;</a></span>ew5</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">integersize* RNA::ew5</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="a835a2d2ed7ac6e253920b916949bea46"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a835a2d2ed7ac6e253920b916949bea46">&#9670;&nbsp;</a></span>ewmb</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">DynProgArray&lt;integersize&gt; * RNA::ewmb</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="a8e958d96bdc5439caac1f1d7471ac0e9"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8e958d96bdc5439caac1f1d7471ac0e9">&#9670;&nbsp;</a></span>ewmb2</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">DynProgArray&lt;integersize&gt; * RNA::ewmb2</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="a4d886f6e92f47962bf2c43f4d24b52bd"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4d886f6e92f47962bf2c43f4d24b52bd">&#9670;&nbsp;</a></span>fce</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">forceclass* RNA::fce</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="adb078ada0b1e8fc227b448e62f5634d3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#adb078ada0b1e8fc227b448e62f5634d3">&#9670;&nbsp;</a></span>lastErrorDetails</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">string RNA::lastErrorDetails</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="a96865031568d528c1e72ca7c77be6417"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a96865031568d528c1e72ca7c77be6417">&#9670;&nbsp;</a></span>lfce</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">bool* RNA::lfce</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="ad7a7258298a689bf42fc43a41a16effb"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad7a7258298a689bf42fc43a41a16effb">&#9670;&nbsp;</a></span>mod</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">bool * RNA::mod</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="a58544d6b4e93569ef51898208be06d37"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a58544d6b4e93569ef51898208be06d37">&#9670;&nbsp;</a></span>partitionfunctionallocated</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">bool RNA::partitionfunctionallocated</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="a3ee830c6817a9a5da8216589b64a7b4d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3ee830c6817a9a5da8216589b64a7b4d">&#9670;&nbsp;</a></span>pfdata</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">pfdatatable* RNA::pfdata</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="a16c4c8714246b97262acb50b4b2aad1d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a16c4c8714246b97262acb50b4b2aad1d">&#9670;&nbsp;</a></span>progress</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">ProgressHandler* RNA::progress</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="a5701fc30b0ed83b60770f84f929e1909"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5701fc30b0ed83b60770f84f929e1909">&#9670;&nbsp;</a></span>Q</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">PFPRECISION RNA::Q</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="a7dbc86bea8ba359871e6b4ad44ccc0a5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7dbc86bea8ba359871e6b4ad44ccc0a5">&#9670;&nbsp;</a></span>structurecoordinates</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">coordinates* RNA::structurecoordinates</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="aec3ea67d1df58e7b3239f8bf03ed0c55"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aec3ea67d1df58e7b3239f8bf03ed0c55">&#9670;&nbsp;</a></span>v</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">DynProgArray&lt;PFPRECISION&gt; * RNA::v</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="acf9f5a2f7a2f2142b0efe2224419f553"></a>
<h2 class="memtitle"><span class="permalink"><a href="#acf9f5a2f7a2f2142b0efe2224419f553">&#9670;&nbsp;</a></span>vmin</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">int RNA::vmin</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="a7c17cadb7eddf9094421a1bd0c7d22cb"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7c17cadb7eddf9094421a1bd0c7d22cb">&#9670;&nbsp;</a></span>w</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">DynProgArray&lt;PFPRECISION&gt;* RNA::w</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="abab25d8889f843589176b4ca2498fd20"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abab25d8889f843589176b4ca2498fd20">&#9670;&nbsp;</a></span>w3</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">PFPRECISION * RNA::w3</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="a11556061f3292c11b40708b7cefb0b35"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a11556061f3292c11b40708b7cefb0b35">&#9670;&nbsp;</a></span>w5</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">PFPRECISION* RNA::w5</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="a22519dbb4cb68fc353dfabc3a249d99c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a22519dbb4cb68fc353dfabc3a249d99c">&#9670;&nbsp;</a></span>wca</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">PFPRECISION ** RNA::wca</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="a208752cb6f6c11e0be87b8ab22038d93"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a208752cb6f6c11e0be87b8ab22038d93">&#9670;&nbsp;</a></span>wcoax</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">DynProgArray&lt;PFPRECISION&gt; * RNA::wcoax</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="aa9df4fb4c33660eac913f67390c862e0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa9df4fb4c33660eac913f67390c862e0">&#9670;&nbsp;</a></span>wl</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">DynProgArray&lt;PFPRECISION&gt; * RNA::wl</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="ac72cbeed4c3618bf170c4afea4ce2d70"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac72cbeed4c3618bf170c4afea4ce2d70">&#9670;&nbsp;</a></span>wlc</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">DynProgArray&lt;PFPRECISION&gt; * RNA::wlc</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="a7fac02aee6bfebae87d44941693ec66d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7fac02aee6bfebae87d44941693ec66d">&#9670;&nbsp;</a></span>wmb</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">DynProgArray&lt;PFPRECISION&gt; * RNA::wmb</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="a2c62121fec224f1c2f6705534a34f91c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2c62121fec224f1c2f6705534a34f91c">&#9670;&nbsp;</a></span>wmbl</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">DynProgArray&lt;PFPRECISION&gt; * RNA::wmbl</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>RNA_class/<a class="el" href="_r_n_a_8h.html">RNA.h</a></li>
<li>RNA_class/<a class="el" href="_r_n_a_8cpp.html">RNA.cpp</a></li>
</ul>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="navelem"><a class="el" href="class_r_n_a.html">RNA</a></li>
    <li class="footer">Generated by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.13 </li>
  </ul>
</div>
</body>
</html>
